<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.5.3" />
<title>rbatools.rba_ModelStructure API documentation</title>
<meta name="description" content="" />
<link href='https://cdnjs.cloudflare.com/ajax/libs/normalize/8.0.0/normalize.min.css' rel='stylesheet'>
<link href='https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/8.0.0/sanitize.min.css' rel='stylesheet'>
<link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css" rel="stylesheet">
<style>.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{font-weight:bold}#index h4 + ul{margin-bottom:.6em}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.name small{font-weight:normal}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase;cursor:pointer}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}.admonition{padding:.1em .5em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title"><code>rbatools.rba_ModelStructure</code> module</h1>
</header>
<section id="section-intro">
<details class="source">
<summary>Source code</summary>
<pre><code class="python"># python 2/3 compatibility
from __future__ import division, print_function

# global imports
import sys
import os.path
import numpy
import pandas
import copy
import json
import jxmlease

# import xml.etree.ElementTree as ET
import libsbml
# package imports
import rba
from rba.core.constraint_blocks import ConstraintBlocks

from rbatools.infoMatrices import InfoMatrices
from rbatools.description_block import DescriptionBlock
from rbatools.metabolite_block import MetaboliteBlock
from rbatools.module_block import ModuleBlock
from rbatools.process_block import ProcessBlock
from rbatools.reaction_block import ReactionBlock
from rbatools.enzyme_block import EnzymeBlock
from rbatools.protein_block import ProteinBlock
from rbatools.macromolecule_block import MacromoleculeBlock
from rbatools.compartment_block import CompartmentBlock
from rbatools.metabolite_constraints import MetaboliteConstraints
from rbatools.density_constraints import DensityConstraints
from rbatools.process_constraints import ProcessConstraints
from rbatools.enzyme_constraints import EnzymeConstraints
from rbatools.statistics_block import StatisticsBlock
from rbatools.target_block import TargetBlock

from sbtab import SBtab


class RBA_ModelStructure(object):
    &#34;&#34;&#34;
    Class holding information on model-structure.

    Attributes
    ----------
    GeneralInfo : rbatools.description_block.Description_block
         Model description
    MetaboliteInfo : rbatools.metabolite_block.Metabolite_block
         Metabolite information
    ModuleInfo : rbatools.module_block.Module_block
         Module information
    ProcessInfo : rbatools.process_block.Process_block
         Process information
    ReactionInfo : rbatools.reaction_block.Reaction_block
         Reaction information
    EnzymeInfo : rbatools.enzyme_block.Enzyme_block
         Enzyme information
    ProteinInfo : rbatools.protein_block.Protein_block
         Protein information
    MacromoleculeInfo : rbatools.macromolecule_block.MacromoleculeBlock
         Macromolecule (other than protein) information
    CompartmentInfo : rbatools.compartment_block.Compartment_block
         Compartment information
    MetaboliteConstraintsInfo : rbatools.metabolite_constraints.Metabolite_constraints
         Metabolite-constraint information
    DensityConstraintsInfo : rbatools.density_constraints.Density_constraints
         Density-constraint information
    ProcessConstraintsInfo : rbatools.process_constraints.Process_constraints
         Process-constraint information
    EnzymeConstraintsInfo : rbatools.enzyme_constraints.Enzyme_constraints
         Enzyme-constraint information
    ModelStatistics : rbatools.statistics_block.Statistics_block
         Statistics on Model
    TargetInfo : rbatools.target_block.TargetBlock
         Target information
    ProteinMatrix : numpy.array
        Matrix mapping proteins to consumers (enzymes and process-machineries)
    MediumDependencies : dict
        Dictionary with boundary metabolites as keys
        and a list of constraint_IDs, which are affected by its concentration, as values.
    MuDependencies : list
        List of constraint_IDs, which are affected by the growth-rate

    Methods
    ----------
    fromFiles:
        Generates model-structure object from model-files and provided auxilliary information.
    fromJSON:
        Generates model-structure object from model-structure in JSON-format.
    exportJSON:
        Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.
    exportSBtab:
        Saves model-structure object in SBtab-format under the specified name (filename).
    generateMatrices:
        Generates information-matricx object and stores it as attribute &#39;InfoMatrices&#39;
    &#34;&#34;&#34;

    def fromFiles(self, xml_dir):
        &#34;&#34;&#34;
        Generates model-structure object from model-files and provided auxilliary information.

        Parameters
        ----------
        xml_dir : str
            Directory, where RBA-model is located
        &#34;&#34;&#34;

        UniprotFile = importUniprotFile(xml_dir)
        GeneMap = importGeneAnnotations(xml_dir)
        Info = importModelInfo(xml_dir)
        SBMLfile = str(&#39;Not There&#39;)
        if Info[&#39;Value&#39;][&#39;SBML-file&#39;] != &#39;Not Provided&#39;:
            SBMLfile = importSbmlFile(xml_dir, str(Info[&#39;Value&#39;][&#39;SBML-file&#39;]))

        MetaboliteAnnotations = importMetaboliteAnnotations(xml_dir)
        ReactionAnnotations = importReactionAnnotations(xml_dir)

        print(&#39;&#39;)
        print(&#39;Generating model-structure&#39;)
        print(&#39;...&#39;)

        model = rba.RbaModel.from_xml(xml_dir)
        Zero_matrix = rba.ConstraintMatrix(model)
        Zero_matrix.build_matrices(0)
        constraints = sortConstraints(Zero_matrix, model)

        MetaboliteInfo = MetaboliteBlock()
        ModuleInfo = ModuleBlock()
        ProcessInfo = ProcessBlock()
        ReactionInfo = ReactionBlock()
        EnzymeInfo = EnzymeBlock()
        ProteinInfo = ProteinBlock()
        MacromoleculeInfo = MacromoleculeBlock()
        CompartmentInfo = CompartmentBlock()
        TargetInfo = TargetBlock()

        TargetInfo.fromFiles(model)
        MetaboliteInfo.fromFiles(model, Info, MetaboliteAnnotations, SBMLfile)
        ModuleInfo.fromFiles(model, SBMLfile)
        ProcessInfo.fromFiles(model, Info)
        ReactionInfo.fromFiles(model, Info, ReactionAnnotations, SBMLfile, MetaboliteInfo)
        EnzymeInfo.fromFiles(model, Info)
        ProteinInfo.fromFiles(model, GeneMap, Info, UniprotFile)
        MacromoleculeInfo.fromFiles(model)
        CompartmentInfo.fromFiles(model, Info)

        self.GeneralInfo = DescriptionBlock()
        self.GeneralInfo.fromFiles(Info)
        self.MetaboliteInfo = MetaboliteInfo
        self.ModuleInfo = ModuleInfo
        self.ProcessInfo = ProcessInfo
        self.TargetInfo = TargetInfo

        for protein in ProteinInfo.Elements.keys():
            AssociatedEnzyme = findAssociatedEnzyme(EnzymeInfo.Elements, protein)
            ProteinInfo.Elements[protein][&#39;associatedEnzymes&#39;] = AssociatedEnzyme[&#39;Enz&#39;]
            ProteinInfo.Elements[protein][&#39;associatedReactions&#39;] = AssociatedEnzyme[&#39;Rx&#39;]

        CB = ConstraintBlocks(model)
        for enzyme in EnzymeInfo.Elements.keys():
            EnzymeInfo.Elements[enzyme][&#39;Isozymes&#39;] = findIsozymes(
                enzyme, CB, ReactionInfo.Elements, EnzymeInfo.Elements[enzyme][&#39;Reaction&#39;])

        for rx in ReactionInfo.Elements.keys():
            if ReactionInfo.Elements[rx][&#39;Enzyme&#39;] is not &#39;&#39;:
                ReactionInfo.Elements[rx][&#39;Compartment_Machinery&#39;] = EnzymeInfo.Elements[ReactionInfo.Elements[rx]
                                                                                         [&#39;Enzyme&#39;]][&#39;EnzymeCompartment&#39;]

        for comp in CompartmentInfo.Elements.keys():
            ContainedMacromolecules = findContainedMacromolecules(comp, MacromoleculeInfo.Elements)
            ContainedProteins = findContainedProteins(comp, ProteinInfo.Elements)
            ContainedEnzymes = findContainedEnzymes(
                ContainedProteins, ProteinInfo.Elements, EnzymeInfo.Elements)
            CompartmentInfo.Elements[comp][&#39;associatedMacromolecules&#39;] = ContainedMacromolecules
            CompartmentInfo.Elements[comp][&#39;associatedProteins&#39;] = ContainedProteins
            CompartmentInfo.Elements[comp][&#39;associatedReactions&#39;] = ContainedEnzymes[&#39;containedReactions&#39;]
            CompartmentInfo.Elements[comp][&#39;associatedEnzymes&#39;] = ContainedEnzymes[&#39;containedEnzymes&#39;]

        self.ReactionInfo = ReactionInfo
        self.EnzymeInfo = EnzymeInfo
        self.ProteinInfo = ProteinInfo
        self.MacromoleculeInfo = MacromoleculeInfo
        self.CompartmentInfo = CompartmentInfo

        self.MetaboliteConstraintsInfo = MetaboliteConstraints()
        self.DensityConstraintsInfo = DensityConstraints()
        self.ProcessConstraintsInfo = ProcessConstraints()
        self.EnzymeConstraintsInfo = EnzymeConstraints()

        self.MetaboliteConstraintsInfo.fromFiles(constraints, Zero_matrix)
        self.DensityConstraintsInfo.fromFiles(model, constraints, Zero_matrix)
        self.ProcessConstraintsInfo.fromFiles(model, constraints, Zero_matrix)
        self.EnzymeConstraintsInfo.fromFiles(model, constraints, Zero_matrix)

        BioConstraintStats = StatsConstraintsBiological(self.MetaboliteConstraintsInfo.Elements,
                                                        self.EnzymeConstraintsInfo.Elements,
                                                        self.DensityConstraintsInfo.Elements,
                                                        self.ProcessConstraintsInfo.Elements)

        MathConstraintStats = StatsConstraintsMathematical(self.MetaboliteConstraintsInfo.Elements,
                                                           self.EnzymeConstraintsInfo.Elements,
                                                           self.DensityConstraintsInfo.Elements,
                                                           self.ProcessConstraintsInfo.Elements,
                                                           self.EnzymeInfo.Elements,
                                                           self.ReactionInfo.Elements,
                                                           self.ProcessInfo.Elements)

        FullOverview = generateOverview(self.ReactionInfo.overview(),
                                        self.MetaboliteInfo.overview(),
                                        self.ModuleInfo.overview(),
                                        self.EnzymeInfo.overview(),
                                        self.ProteinInfo.overview(),
                                        self.MacromoleculeInfo.overview(),
                                        self.ProcessInfo.overview(),
                                        self.CompartmentInfo.overview(),
                                        self.TargetInfo.overview(),
                                        BioConstraintStats,
                                        MathConstraintStats)

        self.ModelStatistics = StatisticsBlock()
        self.ModelStatistics.derive(FullOverview)

        self.ProteinMatrix = generateProteinMatrix(self)
        self.ProteinGeneMatrix = generateProteinGeneMatrix(self)
        self.MediumDependencies, self.MuDependencies = findParameterDependencies(self)

    def fromJSON(self, inputString):
        &#34;&#34;&#34;
        Generates model-structure object from model-structure in JSON-format.

        Parameters
        ----------
        inputString : json-str
            JSON-string to be parsed in to ModelStructure-object.
        &#34;&#34;&#34;
        Block = json.loads(inputString)
        self.ModelStatistics = StatisticsBlock()
        self.GeneralInfo = DescriptionBlock()
        self.ProcessInfo = ProcessBlock()
        self.CompartmentInfo = CompartmentBlock()
        self.MetaboliteInfo = MetaboliteBlock()
        self.TargetInfo = TargetBlock()
        self.ModuleInfo = ModuleBlock()
        self.EnzymeInfo = EnzymeBlock()
        self.ProteinInfo = ProteinBlock()
        self.MacromoleculeInfo = MacromoleculeBlock()
        self.ReactionInfo = ReactionBlock()
        self.DensityConstraintsInfo = DensityConstraints()
        self.ProcessConstraintsInfo = ProcessConstraints()
        self.MetaboliteConstraintsInfo = MetaboliteConstraints()
        self.EnzymeConstraintsInfo = EnzymeConstraints()

        self.ModelStatistics.fromDict(Block[&#39;ModelStatistics&#39;])
        self.GeneralInfo.fromDict(Block[&#39;ModelInformation&#39;])
        self.ProcessInfo.fromDict(Block[&#39;Process&#39;])
        self.CompartmentInfo.fromDict(Block[&#39;Compartment&#39;])
        self.MetaboliteInfo.fromDict(Block[&#39;Metabolite&#39;])
        self.ModuleInfo.fromDict(Block[&#39;Module&#39;])
        self.EnzymeInfo.fromDict(Block[&#39;Enzyme&#39;])
        self.ProteinInfo.fromDict(Block[&#39;Protein&#39;])
        self.MacromoleculeInfo.fromDict(Block[&#39;Macromolecule&#39;])
        self.ReactionInfo.fromDict(Block[&#39;Reaction&#39;])
        self.DensityConstraintsInfo.fromDict(Block[&#39;DensityConstraint&#39;])
        self.ProcessConstraintsInfo.fromDict(Block[&#39;ProcessConstraint&#39;])
        self.MetaboliteConstraintsInfo.fromDict(Block[&#39;MetaboliteConstraint&#39;])
        self.EnzymeConstraintsInfo.fromDict(Block[&#39;EnzymeConstraint&#39;])
        self.TargetInfo.fromDict(Block[&#39;Target&#39;])
        self.ProteinMatrix = Block[&#39;ProteinMatrix&#39;]
        self.ProteinMatrix[&#39;Matrix&#39;] = numpy.array(self.ProteinMatrix[&#39;Matrix&#39;])
        self.ProteinGeneMatrix = Block[&#39;ProteinGeneMatrix&#39;]
        self.ProteinGeneMatrix[&#39;Matrix&#39;] = numpy.array(self.ProteinGeneMatrix[&#39;Matrix&#39;])
        self.MediumDependencies = Block[&#39;MediumDependencies&#39;]
        self.MuDependencies = Block[&#39;MuDependencies&#39;]

    def exportJSON(self, path):
        &#34;&#34;&#34;
        Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.

        Parameters
        ----------
        path : str
            Directory, where to save JSON-file
        &#34;&#34;&#34;
        Block = {&#39;ModelInformation&#39;: self.GeneralInfo.Elements,
                 &#39;ModelStatistics&#39;: self.ModelStatistics.Elements,
                 &#39;Process&#39;: self.ProcessInfo.Elements,
                 &#39;Compartment&#39;: self.CompartmentInfo.Elements,
                 &#39;Metabolite&#39;: self.MetaboliteInfo.Elements,
                 &#39;Target&#39;: self.TargetInfo.Elements,
                 &#39;Module&#39;: self.ModuleInfo.Elements,
                 &#39;Enzyme&#39;: self.EnzymeInfo.Elements,
                 &#39;Protein&#39;: self.ProteinInfo.Elements,
                 &#39;Macromolecule&#39;: self.MacromoleculeInfo.Elements,
                 &#39;Reaction&#39;: self.ReactionInfo.Elements,
                 &#39;DensityConstraint&#39;: self.DensityConstraintsInfo.Elements,
                 &#39;ProcessConstraint&#39;: self.ProcessConstraintsInfo.Elements,
                 &#39;MetaboliteConstraint&#39;: self.MetaboliteConstraintsInfo.Elements,
                 &#39;EnzymeConstraint&#39;: self.EnzymeConstraintsInfo.Elements,
                 &#39;ProteinMatrix&#39;: self.ProteinMatrix,
                 &#39;ProteinGeneMatrix&#39;: self.ProteinGeneMatrix,
                 &#39;MediumDependencies&#39;: self.MediumDependencies,
                 &#39;MuDependencies&#39;: self.MuDependencies}
        # blocks.metabolism.external, model.metabolism.reactions
        # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;S&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
        # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;Reactions&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
        # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;Metabolites&#39;] = [i.id for i in blocks.metabolism.external]
        Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
        Block[&#39;ProteinGeneMatrix&#39;][&#39;Matrix&#39;] = Block[&#39;ProteinGeneMatrix&#39;][&#39;Matrix&#39;].tolist()
        JSONstring = json.dumps(Block, default=JSON_Int64_compensation)
        filename = path + &#39;/ModelStructure.json&#39;
        f = open(filename, &#39;w&#39;)
        f.write(JSONstring)
        f.close()
        return(JSONstring)

    def exportSBtab(self, add_links=False, filename=None):
        &#34;&#34;&#34;
        Saves model-structure object in SBtab-format under the specified name (filename).

        Parameters
        ----------
        filename : str
            Name, under which to save SBtab-file
        add_links : str
            Wheter to implement entry-format, which allows links between table-elements.
        &#34;&#34;&#34;
        GeneralInfoTable = self.GeneralInfo.toSBtab(table_id=&#39;ModelMetadata&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Model Information&#39;, Col_list=[
                                                    &#39;Measure&#39;, &#39;Value&#39;], NameList=[&#39;Element&#39;, &#39;Value&#39;])
        GeneralInfoTable.filename = &#39;ModelMetadata.tsv&#39;
        GeneralInfoTable.change_attribute(&#39;Text&#39;, &#39;Model metadata&#39;)
        GeneralInfoTable.unset_attribute(&#39;Date&#39;)
        GeneralInfoTable.unset_attribute(&#39;SBtabVersion&#39;)

        StatsTable = self.ModelStatistics.toSBtab(
            table_id=&#39;ModelElements&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Model Size&#39;, Col_list=[&#39;Measure&#39;, &#39;Value&#39;], NameList=[&#39;Element&#39;, &#39;Number&#39;])
        StatsTable.filename = &#39;ModelElements.tsv&#39;
        StatsTable.change_attribute(&#39;Text&#39;, &#39;Numbers of cell elements and constraints&#39;)
        StatsTable.unset_attribute(&#39;Date&#39;)
        StatsTable.unset_attribute(&#39;SBtabVersion&#39;)

        MetaboliteBlock_forChanges = copy.deepcopy(self.MetaboliteInfo)
        if add_links:
            for k in list(MetaboliteBlock_forChanges.Elements.keys()):
                if MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;] in self.CompartmentInfo.Elements.keys():
                    MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + &#39;Compartment&#39; + \
                        &#39;/&#39;+MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;]+&#39;!)&#39;
                for index, id in enumerate(MetaboliteBlock_forChanges.Elements[k][&#39;ReactionsInvolvedWith&#39;]):
                    MetaboliteBlock_forChanges.Elements[k][&#39;ReactionsInvolvedWith&#39;][
                        index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for id in list(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                    if not pandas.isna(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id]):
                        MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id] = str(&#39;(!identifiers:&#39;+id+&#39;/&#39;+str(
                            MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;|&#39;+str(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;!)&#39;)

        MetaboliteTable = MetaboliteBlock_forChanges.toSBtab(table_id=&#39;Compound&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Metabolites&#39;, Col_list=[
            &#39;ID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;ReactionsInvolvedWith&#39;, &#39;OtherIDs&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;Reactions&#39;, &#39;Annotation&#39;])
        MetaboliteTable.filename = &#39;Compound.tsv&#39;
        MetaboliteTable.change_attribute(
            &#39;Text&#39;, &#39;Metabolite species are localised in cell compartments and are associated with metabolite mass balance constraints.&#39;)
        MetaboliteTable.unset_attribute(&#39;Date&#39;)
        MetaboliteTable.unset_attribute(&#39;SBtabVersion&#39;)

        ReactionBlock_forChanges = copy.deepcopy(self.ReactionInfo)
        if add_links:
            for k in list(ReactionBlock_forChanges.Elements.keys()):
                oldEnz = ReactionBlock_forChanges.Elements[k][&#39;Enzyme&#39;]
                if len(oldEnz) &gt; 0:
                    ReactionBlock_forChanges.Elements[k][&#39;Enzyme&#39;] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+oldEnz+&#39;!)&#39;

                replacements = {}
                for i in ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;].split(&#39; &#39;):
                    if i in self.MetaboliteInfo.Elements.keys():
                        replacements[i] = &#39;(!&#39;+&#39;Compound&#39; + &#39;/&#39;+i+&#39;!)&#39;
                for i in replacements.keys():
                    ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;].replace(
                        str(&#39; &#39;+i+&#39; &#39;), str(&#39; &#39;+replacements[i]+&#39; &#39;))

                for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Compartment_Species&#39;]):
                    if id in self.CompartmentInfo.Elements.keys():
                        ReactionBlock_forChanges.Elements[k][&#39;Compartment_Species&#39;][
                            index] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;

                for id in list(ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;].keys()):
                    ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;][&#39;(!&#39;+&#39;Compound&#39; +
                                                                      &#39;/&#39;+id+&#39;!)&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;].pop(id)
                for id in list(ReactionBlock_forChanges.Elements[k][&#39;Products&#39;].keys()):
                    ReactionBlock_forChanges.Elements[k][&#39;Products&#39;][&#39;(!&#39;+&#39;Compound&#39; +
                                                                     &#39;/&#39;+id+&#39;!)&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Products&#39;].pop(id)
                for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Twins&#39;]):
                    ReactionBlock_forChanges.Elements[k][&#39;Twins&#39;][index] = &#39;(!&#39; + \
                        &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Compartment_Machinery&#39;]):
                    ReactionBlock_forChanges.Elements[k][&#39;Compartment_Machinery&#39;][
                        index] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;
                if &#39;ProtoID&#39; in list(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                    ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].pop(&#39;ProtoID&#39;)
                for id in list(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                    if not pandas.isna(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id]):
                        ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id] = str(&#39;(!identifiers:&#39;+id+&#39;/&#39;+str(
                            ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;|&#39;+str(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;!)&#39;)

        ReactionTable = ReactionBlock_forChanges.toSBtab(table_id=&#39;Reaction&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Reactions&#39;, Col_list=[&#39;ID&#39;, &#39;Name&#39;, &#39;Type&#39;, &#39;Reversible&#39;, &#39;Formula&#39;, &#39;Enzyme&#39;, &#39;Compartment_Machinery&#39;, &#39;Twins&#39;, &#39;Compartment_Species&#39;, &#39;OtherIDs&#39;], NameList=[
            &#39;ID&#39;, &#39;Name&#39;, &#39;Type&#39;, &#39;IsReversible&#39;, &#39;ReactionFormula&#39;, &#39;Enzyme&#39;, &#39;EnzymeCompartment&#39;, &#39;IsoenzymeReactions&#39;, &#39;ReactionCompartment&#39;, &#39;Annotation&#39;])
        ReactionTable.filename = &#39;Reaction.tsv&#39;
        ReactionTable.change_attribute(
            &#39;Text&#39;, &#39;Chemical reactions are localised in cell compartments. All reactions are enzyme catalysed.&#39;)
        ReactionTable.unset_attribute(&#39;Date&#39;)
        ReactionTable.unset_attribute(&#39;SBtabVersion&#39;)

        EnzymeBlock_forChanges = copy.deepcopy(self.EnzymeInfo)
        if add_links:
            for k in list(EnzymeBlock_forChanges.Elements.keys()):
                oldRx = EnzymeBlock_forChanges.Elements[k][&#39;Reaction&#39;]
                EnzymeBlock_forChanges.Elements[k][&#39;Reaction&#39;] = &#39;(!&#39;+&#39;Reaction&#39;+&#39;/&#39;+oldRx+&#39;!)&#39;
                for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;Isozymes&#39;]):
                    EnzymeBlock_forChanges.Elements[k][&#39;Isozymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;IdenticalEnzymes&#39;]):
                    EnzymeBlock_forChanges.Elements[k][&#39;IdenticalEnzymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;EnzymeCompartment&#39;]):
                    EnzymeBlock_forChanges.Elements[k][&#39;EnzymeCompartment&#39;][index] = &#39;(!&#39; + \
                        &#39;Compartment&#39;+&#39;/&#39; + id+&#39;!)&#39;
                for id in list(EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;].keys()):
                    EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;][&#39;(!&#39;+&#39;Protein&#39;+&#39;/&#39; +
                                                                   id+&#39;!)&#39;] = EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;].pop(id)

        EnzymeTable = EnzymeBlock_forChanges.toSBtab(table_id=&#39;Enzyme&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Enzymes&#39;, Col_list=[
            &#39;ID&#39;, &#39;Reaction&#39;, &#39;IdenticalEnzymes&#39;, &#39;Subunits&#39;, &#39;EnzymeCompartment&#39;, &#39;OtherIDs&#39;, &#39;Isozymes&#39;], NameList=[&#39;ID&#39;, &#39;CatalysedReaction&#39;, &#39;OtherEnzymaticActivities&#39;, &#39;Subunits&#39;, &#39;Compartment&#39;, &#39;Annotation&#39;, &#39;Isoenzymes&#39;])
        EnzymeTable.filename = &#39;Enzyme.tsv&#39;
        EnzymeTable.change_attribute(
            &#39;Text&#39;, &#39;Enzymes are localised in cell compartments and catalyse specific reactions. To describe multi-functional enzyme, RBA uses multiple enzyme entries. Enzymes are associated with enzyme capacity constraints.&#39;)
        EnzymeTable.unset_attribute(&#39;Date&#39;)
        EnzymeTable.unset_attribute(&#39;SBtabVersion&#39;)

        ProteinBlock_forChanges = copy.deepcopy(self.ProteinInfo)
        if add_links:
            for k in list(ProteinBlock_forChanges.Elements.keys()):
                oldComp = ProteinBlock_forChanges.Elements[k][&#39;Compartment&#39;]
                ProteinBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + \
                    &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;
                for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;associatedReactions&#39;]):
                    ProteinBlock_forChanges.Elements[k][&#39;associatedReactions&#39;][
                        index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;]):
                    ProteinBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;]):
                    ProteinBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;][index] = &#39;(!&#39; + \
                        &#39;Process&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for id in list(ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].keys()):
                    ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;][
                        &#39;(!&#39;+&#39;Process&#39;+&#39;/&#39; + id+&#39;!)&#39;] = ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].pop(id)
                for id in list(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;].keys()):
                    if not pandas.isna(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id]):
                        if id == &#39;UniprotID&#39;:
                            ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] = str(&#39;(!identifiers:uniprot/&#39;+str(
                                ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;|&#39;+str(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;!)&#39;)
                        if id == &#39;ECnumber&#39;:
                            if ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] != &#39;nan&#39;:
                                ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] = str(&#39;(!identifiers:ec-code/&#39;+str(
                                    ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id]).split(&#39;EC &#39;)[1]+&#39;|&#39;+str(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;!)&#39;)

        ProteinTable = ProteinBlock_forChanges.toSBtab(table_id=&#39;Protein&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Proteins&#39;, Col_list=[&#39;ID&#39;, &#39;ProtoID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;SupportsProcess&#39;, &#39;associatedEnzymes&#39;, &#39;associatedReactions&#39;, &#39;AAnumber&#39;, &#39;ProcessRequirements&#39;, &#39;Weight&#39;, &#39;ExternalIDs&#39;, &#39;Function&#39;], NameList=[
            &#39;ID&#39;, &#39;ProtoID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;ContributesToProcess&#39;, &#39;EnzymaticActivity&#39;, &#39;CatalysedReaction&#39;, &#39;ChainLength&#39;, &#39;RequiresProcess&#39;, &#39;MolecularWeight&#39;, &#39;Annotation&#39;, &#39;Function&#39;])
        ProteinTable.filename = &#39;Protein.tsv&#39;
        ProteinTable.change_attribute(
            &#39;Text&#39;, &#39;Proteins are localised in cell compartments and can be (or be part of) metabolic enzymes.&#39;)
        ProteinTable.unset_attribute(&#39;Date&#39;)
        ProteinTable.unset_attribute(&#39;SBtabVersion&#39;)

        MacromoleculeBlock_forChanges = copy.deepcopy(self.MacromoleculeInfo)
        if add_links:
            for k in list(MacromoleculeBlock_forChanges.Elements.keys()):
                oldComp = MacromoleculeBlock_forChanges.Elements[k][&#39;Compartment&#39;]
                MacromoleculeBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + \
                    &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;
                for index, id in enumerate(MacromoleculeBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;]):
                    MacromoleculeBlock_forChanges.Elements[k][
                        &#39;SupportsProcess&#39;][index] = &#39;(!&#39; + &#39;Process&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for id in list(MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].keys()):
                    MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;][
                        &#39;(!&#39;+&#39;Process&#39;+&#39;/&#39; + id+&#39;!)&#39;] = MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].pop(id)

        MacromoleculeTable = MacromoleculeBlock_forChanges.toSBtab(table_id=&#39;Macromolecule&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Macromolecules&#39;, Col_list=[
                                                                   &#39;ID&#39;, &#39;ProtoID&#39;, &#39;Type&#39;, &#39;Compartment&#39;, &#39;SupportsProcess&#39;, &#39;ProcessRequirements&#39;], NameList=[&#39;ID&#39;, &#39;ProtoID&#39;, &#39;Type&#39;, &#39;Compartment&#39;, &#39;SupportedtedProcess&#39;, &#39;RequiresProcess&#39;])
        MacromoleculeTable.filename = &#39;Macromolecules.tsv&#39;
        MacromoleculeTable.change_attribute(
            &#39;Text&#39;, &#39;Macromolecule are localised in cell compartments and can be part of cellular machinery or serve another function inside the cell.&#39;)
        MacromoleculeTable.unset_attribute(&#39;Date&#39;)
        MacromoleculeTable.unset_attribute(&#39;SBtabVersion&#39;)

        CompartmentBlock_forChanges = copy.deepcopy(self.CompartmentInfo)
        if add_links:
            for k in list(CompartmentBlock_forChanges.Elements.keys()):
                for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedReactions&#39;]):
                    CompartmentBlock_forChanges.Elements[k][&#39;associatedReactions&#39;][
                        index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;]):
                    CompartmentBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedProteins&#39;]):
                    CompartmentBlock_forChanges.Elements[k][&#39;associatedProteins&#39;][
                        index] = &#39;(!&#39; + &#39;Protein&#39;+&#39;/&#39;+id+&#39;!)&#39;

        CompartmentTable = CompartmentBlock_forChanges.toSBtab(table_id=&#39;Compartment&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Compartments&#39;, Col_list=[
                                                               &#39;ID&#39;, &#39;associatedProteins&#39;, &#39;associatedEnzymes&#39;, &#39;associatedReactions&#39;], NameList=[&#39;ID&#39;, &#39;Proteins&#39;, &#39;Enzymes&#39;, &#39;Reactions&#39;])

        CompartmentTable.filename = &#39;Compartment.tsv&#39;
        CompartmentTable.change_attribute(
            &#39;Text&#39;, &#39;Cell compartments are used to describe the localisation of proteins, enzymes, and reactions and are associated with density constraints.&#39;)
        CompartmentTable.unset_attribute(&#39;Date&#39;)
        CompartmentTable.unset_attribute(&#39;SBtabVersion&#39;)

        ModuleBlock_forChanges = copy.deepcopy(self.ModuleInfo)
        if add_links:
            for k in list(ModuleBlock_forChanges.Elements.keys()):
                for index, id in enumerate(ModuleBlock_forChanges.Elements[k][&#39;Members&#39;]):
                    if id in self.ReactionInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                    if id in self.ProteinInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Protein&#39;+&#39;/&#39;+id+&#39;!)&#39;
                    if id in self.MetaboliteInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Compound&#39;+&#39;/&#39;+id+&#39;!)&#39;
                    if id in self.CompartmentInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;

        ModuleTable = ModuleBlock_forChanges.toSBtab(table_id=&#39;CellModule&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Modules&#39;, Col_list=[
                                                     &#39;ID&#39;, &#39;Name&#39;, &#39;Members&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;Contains&#39;])
        ModuleTable.filename = &#39;Module.tsv&#39;
        ModuleTable.change_attribute(&#39;Text&#39;, &#39;Information on Modules in RBAmodel&#39;)
        ModuleTable.unset_attribute(&#39;Date&#39;)
        ModuleTable.unset_attribute(&#39;SBtabVersion&#39;)

        ProcessBlock_forChanges = copy.deepcopy(self.ProcessInfo)
        if add_links:
            for k in list(ProcessBlock_forChanges.Elements.keys()):
                for component in ProcessBlock_forChanges.Elements[k][&#39;Components&#39;].keys():
                    ProductKeys = list(
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;].keys())
                    for species in ProductKeys:
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;][
                            &#39;(!&#39;+&#39;Compound&#39;+&#39;/&#39;+species+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;].pop(species)
                    ReactantKeys = list(
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;].keys())
                    for species in ReactantKeys:
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;][
                            &#39;(!&#39;+&#39;Compound&#39;+&#39;/&#39;+species+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;].pop(species)
                for id in list(ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].keys()):
                    if id in self.ProteinInfo.Elements.keys():
                        ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Protein&#39;+&#39;/&#39; +
                                                                           id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)
                    elif id in self.MacromoleculeInfo.Elements.keys():
                        ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Macromolecule&#39; +
                                                                           &#39;/&#39; + id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)
                    elif id in self.MetaboliteInfo.Elements.keys():
                        ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Metabolite&#39;+&#39;/&#39; +
                                                                           id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)

        ProcessTable = ProcessBlock_forChanges.toSBtab(table_id=&#39;Process&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Macromolecular Processes&#39;, Col_list=[
            &#39;ID&#39;, &#39;Name&#39;, &#39;Composition&#39;, &#39;Components&#39;, &#39;Initiation&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;MachineSubunits&#39;, &#39;MachineComponents&#39;, &#39;InitiationCofactors&#39;])
        ProcessTable.filename = &#39;Process.tsv&#39;
        ProcessTable.change_attribute(
            &#39;Text&#39;, &#39;Macromolecular machines catalyse the biochemical reactions that produce, modify, and degrade macromolecules. They are catalysed by macromolecular machines and associated with process capacity constraints.&#39;)
        ProcessTable.unset_attribute(&#39;Date&#39;)
        ProcessTable.unset_attribute(&#39;SBtabVersion&#39;)

        MetaboliteConstraintsBlock_forChanges = copy.deepcopy(self.MetaboliteConstraintsInfo)
        if add_links:
            for k in list(MetaboliteConstraintsBlock_forChanges.Elements.keys()):
                oldMet = MetaboliteConstraintsBlock_forChanges.Elements[k][&#39;AssociatedMetabolite&#39;]
                MetaboliteConstraintsBlock_forChanges.Elements[k][
                    &#39;AssociatedMetabolite&#39;] = &#39;(!&#39; + &#39;Compound&#39;+&#39;/&#39;+oldMet+&#39;!)&#39;

        MetaboliteConstraintTable = MetaboliteConstraintsBlock_forChanges.toSBtab(table_id=&#39;MetaboliteConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Metabolite Mass Balance Constraints&#39;, Col_list=[
                                                                                  &#39;ID&#39;, &#39;AssociatedMetabolite&#39;, &#39;Type&#39;], NameList=[&#39;ID&#39;, &#39;Metabolite&#39;, &#39;Type&#39;])
        MetaboliteConstraintTable.filename = &#39;MetaboliteConstraint.tsv&#39;
        MetaboliteConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;Metabolite mass balance constraints ensure mass balance and stationary fluxes in metabolism.&#39;)
        MetaboliteConstraintTable.unset_attribute(&#39;Date&#39;)
        MetaboliteConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        CompartmentConstraintsBlock_forChanges = copy.deepcopy(self.DensityConstraintsInfo)
        if add_links:
            for k in list(CompartmentConstraintsBlock_forChanges.Elements.keys()):
                oldComp = CompartmentConstraintsBlock_forChanges.Elements[k][&#39;AssociatedCompartment&#39;]
                CompartmentConstraintsBlock_forChanges.Elements[k][
                    &#39;AssociatedCompartment&#39;] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

        DensityConstraintTable = CompartmentConstraintsBlock_forChanges.toSBtab(table_id=&#39;DensityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Density Constraints&#39;, Col_list=[
                                                                                &#39;ID&#39;, &#39;AssociatedCompartment&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;Formula&#39;])
        DensityConstraintTable.filename = &#39;DensityConstraint.tsv&#39;
        DensityConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;Density constraints put an upper bound on the sum of macromolecule concentrations in a given compartment. The capacity parameter defines this bound in units corresponding to amino acids (contained in proteins), or one third of nucleotides (contained in RNA).&#39;)
        DensityConstraintTable.unset_attribute(&#39;Date&#39;)
        DensityConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        ProcessConstraintsBlock_forChanges = copy.deepcopy(self.ProcessConstraintsInfo)
        if add_links:
            for k in list(ProcessConstraintsBlock_forChanges.Elements.keys()):
                oldComp = ProcessConstraintsBlock_forChanges.Elements[k][&#39;AssociatedProcess&#39;]
                ProcessConstraintsBlock_forChanges.Elements[k][&#39;AssociatedProcess&#39;] = &#39;(!&#39; + \
                    &#39;Process&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

        ProcessConstraintTable = ProcessConstraintsBlock_forChanges.toSBtab(table_id=&#39;MachineryCapacityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Machinery Capacity Constraints&#39;, Col_list=[
            &#39;ID&#39;, &#39;AssociatedProcess&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Process&#39;, &#39;Type&#39;, &#39;Formula&#39;])
        ProcessConstraintTable.filename = &#39;MachineryConstraint.tsv&#39;
        ProcessConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;A machinery capacity constraint states that the rate of a macromolecular process is proportional to the concentration of the catalysing machine. The proportionality constant  (capacity parameter) can be a constant or a function of model parameters such as the growth rate.&#39;)
        ProcessConstraintTable.unset_attribute(&#39;Date&#39;)
        ProcessConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        EnzymeConstraintsBlock_forChanges = copy.deepcopy(self.EnzymeConstraintsInfo)
        if add_links:
            for k in list(EnzymeConstraintsBlock_forChanges.Elements.keys()):
                oldComp = EnzymeConstraintsBlock_forChanges.Elements[k][&#39;AssociatedEnzyme&#39;]
                EnzymeConstraintsBlock_forChanges.Elements[k][&#39;AssociatedEnzyme&#39;] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

        EnzymeConstraintTable = EnzymeConstraintsBlock_forChanges.toSBtab(table_id=&#39;EnzymeCapacityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Enzyme Capacity Constraints&#39;, Col_list=[
            &#39;ID&#39;, &#39;AssociatedEnzyme&#39;, &#39;AssociatedReaction&#39;, &#39;Direction&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Enzyme&#39;, &#39;Reaction&#39;, &#39;Direction&#39;, &#39;Type&#39;, &#39;Formula&#39;])
        EnzymeConstraintTable.filename = &#39;EnzymeConstraint.tsv&#39;
        EnzymeConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;An enzyme capacity constraint states that a reaction rate is proportional to the concentration of the catalysing enzyme. The proportionality constant (capacity parameter) can be a constant or a function of model parameters such as the growth rate.&#39;)
        EnzymeConstraintTable.unset_attribute(&#39;Date&#39;)
        EnzymeConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        TargetBlock_forChanges = copy.deepcopy(self.TargetInfo)
        if add_links:
            for k in list(TargetBlock_forChanges.Elements.keys()):
                oldTargSpec = TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;]
                if oldTargSpec in list(self.MetaboliteInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Compound&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
                if oldTargSpec in list(self.ReactionInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Reaction&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
                if oldTargSpec in list(self.ProteinInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Protein&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
                if oldTargSpec in list(self.MacromoleculeInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Macromolecule&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;

        TargetTable = TargetBlock_forChanges.toSBtab(table_id=&#39;CellTarget&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Targets&#39;, Col_list=[
            &#39;ID&#39;, &#39;Group&#39;, &#39;Type&#39;, &#39;TargetEntity&#39;, &#39;TargetValue&#39;], NameList=[&#39;ID&#39;, &#39;TargetGroup&#39;, &#39;TargetType&#39;, &#39;TargetEntity&#39;, &#39;TargetParameter&#39;])
        TargetTable.filename = &#39;Target.tsv&#39;
        TargetTable.change_attribute(
            &#39;Text&#39;, &#39;Cell targets are biochemical variables (fluxes or concentrations) that need to remain below or above a certain threshold value to ensure a viable cell. They define constraints in the RBA model.&#39;)
        TargetTable.unset_attribute(&#39;Date&#39;)
        TargetTable.unset_attribute(&#39;SBtabVersion&#39;)

        if filename is not None:
            filename_SBtab = filename
        else:
            if add_links:
                filename_SBtab = &#39;RBA_model_withLinks&#39;
            else:
                filename_SBtab = &#39;RBA_model&#39;

        Out = SBtab.SBtabDocument(name=&#39;rbatools_withLinks&#39;, sbtab_init=None,
                                  filename=str(filename_SBtab+&#39;.tsv&#39;))
        Out.add_sbtab(GeneralInfoTable)
        Out.add_sbtab(StatsTable)
        Out.add_sbtab(MetaboliteTable)
        Out.add_sbtab(ReactionTable)
        Out.add_sbtab(EnzymeTable)
        Out.add_sbtab(ProteinTable)
        Out.add_sbtab(MacromoleculeTable)
        Out.add_sbtab(CompartmentTable)
        Out.add_sbtab(ModuleTable)
        Out.add_sbtab(ProcessTable)
        Out.add_sbtab(TargetTable)
        Out.add_sbtab(MetaboliteConstraintTable)
        Out.add_sbtab(DensityConstraintTable)
        Out.add_sbtab(ProcessConstraintTable)
        Out.add_sbtab(EnzymeConstraintTable)

        Out.name = filename_SBtab
        Out.change_attribute(&#39;DocumentName&#39;, self.GeneralInfo.Elements[&#39;Name&#39;])
        Out.change_attribute(&#39;DocumentType&#39;, &#39;rba-model-synopsis&#39;)
        Out.write()

    def generateMatrices(self):
        &#34;&#34;&#34;
        Generates information-matricx object and stores it as attribute &#39;InfoMatrices&#39;
        &#34;&#34;&#34;
        self.InfoMatrices = InfoMatrices(self)


def findParameterDependencies(ModelStructure):
    MedDepts = {}
    MuDepts = []
    for eK in ModelStructure.EnzymeConstraintsInfo.Elements.keys():
        e = ModelStructure.EnzymeConstraintsInfo.Elements[eK]
        for pf in e[&#39;CapacityParameter&#39;]:
            iV = list(pf.values())[0][&#39;IndependentVariable&#39;]
            if list(pf.values())[0][&#39;FunctionType&#39;] != &#39;constant&#39;:
                if iV.startswith(&#39;M_&#39;):
                    if iV.rsplit(&#39;_&#39;, 1)[0] in MedDepts.keys():
                        MedDepts[iV.rsplit(&#39;_&#39;, 1)[0]].append(e[&#39;ID&#39;])
                    elif iV.rsplit(&#39;_&#39;, 1)[0] not in MedDepts.keys():
                        MedDepts.update({iV.rsplit(&#39;_&#39;, 1)[0]: [e[&#39;ID&#39;]]})
                if iV == &#39;growth_rate&#39;:
                    if e[&#39;ID&#39;] not in MuDepts:
                        MuDepts.append(e[&#39;ID&#39;])
    for dK in ModelStructure.DensityConstraintsInfo.Elements.keys():
        d = ModelStructure.DensityConstraintsInfo.Elements[dK]
        for pf in d[&#39;CapacityParameter&#39;]:
            iV = list(pf.values())[0][&#39;IndependentVariable&#39;]
            if list(pf.values())[0][&#39;FunctionType&#39;] != &#39;constant&#39;:
                if iV.startswith(&#39;M_&#39;):
                    if iV.rsplit(&#39;_&#39;, 1)[0] in MedDepts.keys():
                        MedDepts[iV.rsplit(&#39;_&#39;, 1)[0]].append(d[&#39;ID&#39;])
                    elif iV.rsplit(&#39;_&#39;, 1)[0] not in MedDepts.keys():
                        MedDepts.update({iV.rsplit(&#39;_&#39;, 1)[0]: [d[&#39;ID&#39;]]})
                if iV == &#39;growth_rate&#39;:
                    if d[&#39;ID&#39;] not in MuDepts:
                        MuDepts.append(d[&#39;ID&#39;])
    for pK in ModelStructure.ProcessConstraintsInfo.Elements.keys():
        p = ModelStructure.ProcessConstraintsInfo.Elements[pK]
        for pf in p[&#39;CapacityParameter&#39;]:
            iV = list(pf.values())[0][&#39;IndependentVariable&#39;]
            if list(pf.values())[0][&#39;FunctionType&#39;] != &#39;constant&#39;:
                if iV.startswith(&#39;M_&#39;):
                    if iV.rsplit(&#39;_&#39;, 1)[0] in MedDepts.keys():
                        MedDepts[iV.rsplit(&#39;_&#39;, 1)[0]].append(p[&#39;ID&#39;])
                    elif iV.rsplit(&#39;_&#39;, 1)[0] not in MedDepts.keys():
                        MedDepts.update({iV.rsplit(&#39;_&#39;, 1)[0]: [p[&#39;ID&#39;]]})
                if iV == &#39;growth_rate&#39;:
                    if p[&#39;ID&#39;] not in MuDepts:
                        MuDepts.append(p[&#39;ID&#39;])
    return(MedDepts, MuDepts)


def generateProteinGeneMatrix(ModelStructure):
    uniqueProteins = []
    uniqueProteinMap = {}
    Proteins = list(ModelStructure.ProteinInfo.Elements.keys())
    for i in ModelStructure.ProteinInfo.Elements.keys():
        if ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;] not in list(uniqueProteinMap.keys()):
            uniqueProteinMap.update({ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;]: []})
            uniqueProteins.append(ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;])
        uniqueProteinMap[ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;]].append(i)
    ProteinProteinMatrix = numpy.zeros(
        (len(list(uniqueProteinMap.keys())), len(list(ModelStructure.ProteinInfo.Elements.keys()))))
    for u in list(uniqueProteinMap.keys()):
        row_ind = uniqueProteins.index(u)
        for i in uniqueProteinMap[u]:
            col_ind = Proteins.index(i)
            ProteinProteinMatrix[row_ind, col_ind] = 1
    return({&#39;Matrix&#39;: numpy.array(ProteinProteinMatrix), &#39;Proteins&#39;: Proteins, &#39;ProtoProteins&#39;: uniqueProteins})


def generateProteinMatrix(ModelStructure):
    Proteins = list(ModelStructure.ProteinInfo.Elements.keys())
    # print(list(ModelStructure.ProcessInfo.Elements.keys()))
    Processes = [ModelStructure.ProcessInfo.Elements[i][&#39;ID&#39;] +
                 &#39;_machinery&#39; for i in list(ModelStructure.ProcessInfo.Elements.keys())]
    Enzymes = list(ModelStructure.EnzymeInfo.Elements.keys())
    Consumers = list(set(list(Enzymes+Processes)))
    ProteinMatrix = numpy.zeros((len(Proteins), len(Consumers)))
    for p in Proteins:
        if len(ModelStructure.ProteinInfo.Elements[p][&#39;SupportsProcess&#39;]) &gt; 0:
            # print(list(ModelStructure.ProteinInfo.Elements[p][&#39;SupportsProcess&#39;]))
            for pc in list(ModelStructure.ProteinInfo.Elements[p][&#39;SupportsProcess&#39;]):
                coeff = 0
                row_ind = Proteins.index(p)
                col_ind = Consumers.index(
                    ModelStructure.ProcessInfo.Elements[pc][&#39;ID&#39;]+&#39;_machinery&#39;)
                coeff = ModelStructure.ProcessInfo.Elements[pc][&#39;Composition&#39;][p]
                ProteinMatrix[row_ind, col_ind] += coeff
        if len(ModelStructure.ProteinInfo.Elements[p][&#39;associatedEnzymes&#39;]) &gt; 0:
            for ez in list(ModelStructure.ProteinInfo.Elements[p][&#39;associatedEnzymes&#39;]):
                coeff = 0
                row_ind = Proteins.index(p)
                col_ind = Consumers.index(ez)
                coeff = ModelStructure.EnzymeInfo.Elements[ez][&#39;Subunits&#39;][p]
                ProteinMatrix[row_ind, col_ind] += coeff
    return({&#39;Matrix&#39;: numpy.array(ProteinMatrix), &#39;Consumers&#39;: Consumers, &#39;Proteins&#39;: Proteins})


def importBiggMetabolites(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/bigg_models_metabolites.txt&#39;)):
        return(pandas.read_csv(str(xml_dir+&#39;/bigg_models_metabolites.txt&#39;), sep=&#39;\t&#39;, index_col=0))
    else:
        sys.exit(&#39;\n Required BiGG Metabolite File &#34;bigg_models_metabolites.txt&#34; not found.\n&#39; +
                 &#39; Please provide in input-directory\n&#39; +
                 &#39; To be found under: http://bigg.ucsd.edu/static/namespace/bigg_models_metabolites.txt\n&#39;)


def importBiggReactions(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/bigg_models_reactions.txt&#39;)):
        return(pandas.read_csv(str(xml_dir+&#39;/bigg_models_reactions.txt&#39;), sep=&#39;\t&#39;, index_col=0))
    else:
        sys.exit(&#39;\n Required BiGG Reaction File &#34;bigg_models_reactions.txt &#34;not found.\n&#39; +
                 &#39; Please provide in input-directory\n&#39; +
                 &#39; To be found under: http://bigg.ucsd.edu/static/namespace/bigg_models_reactions.txt\n&#39;)


def importUniprotFile(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/uniprot.csv&#39;)):
        return(pandas.read_csv(str(xml_dir+&#39;/uniprot.csv&#39;), sep=&#39;\t&#39;))
    else:
        print(&#39;\n Uniprot-file &#34;uniprot.csv&#34; not found.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))


def importSbmlFile(xml_dir, filename):
    if os.path.isfile(str(xml_dir+&#39;/&#39;+filename)):
        SBfile = libsbml.readSBML(str(xml_dir+&#39;/&#39;+filename))
        if SBfile.getNumErrors() &gt; 0:
            SBfile.printErrors()
            print(&#39;Invalid SBML&#39;)
            return(str(&#39;Not There&#39;))
        else:
            sbml = SBfile
            return(sbml)
    else:
        print(&#39;\n SBML-file {} not found.\n&#39; +
              &#39; Continuing without additional information...\n&#39;.format(filename))
        return(str(&#39;Not There&#39;))


def importGeneAnnotations(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/GeneAnnotations.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/GeneAnnotations.csv&#39;), sep=&#39;,&#39;, index_col=0)
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;GeneAnnotations.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No Gene-annotation file &#34;GeneAnnotations.csv&#34; provided.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))


def importReactionAnnotations(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/ReactionAnnotations.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/ReactionAnnotations.csv&#39;), sep=&#39;,&#39;, index_col=0)
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;ReactionAnnotations.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No Reaction-annotation file &#34;ReactionAnnotations.csv&#34; provided.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))


def importMetaboliteAnnotations(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/MetaboliteAnnotations.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/MetaboliteAnnotations.csv&#39;), sep=&#39;,&#39;, index_col=0)
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;MetaboliteAnnotations.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No Reaction-annotation file &#34;MetaboliteAnnotations.csv&#34; provided.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))


def importModelInfo(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/ModelInformation.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/ModelInformation.csv&#39;),
                              sep=&#39;,&#39;, header=0)
        out.index = list(out[&#39;Key&#39;])
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;ModelInformation.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No model-info file &#34;ModelInformation.csv&#34; provided.\n&#39; +
              &#39; Using dummy-information\n&#39;)
        return(pandas.DataFrame([[&#39;Name&#39;, &#39;ModelName&#39;], [&#39;Author&#39;, &#39;John Doe&#39;], [&#39;Organism&#39;, &#39;Life&#39;], [&#39;Reconstruction&#39;, &#39;GSMM&#39;], [&#39;SBML-file&#39;, &#39;Not Provided&#39;]], index=[&#39;Name&#39;, &#39;Author&#39;, &#39;Organism&#39;, &#39;Reconstruction&#39;, &#39;SBML-file&#39;], columns=[&#39;Key&#39;, &#39;Value&#39;]))


def htmlStyle(structOriginal):
    struct = copy.deepcopy(structOriginal)

    for j in struct.ProcessInfo.Elements.keys():
        for i in struct.ProcessInfo.Elements[j][&#39;Composition&#39;].keys():
            struct.ProcessInfo.Elements[j][&#39;Composition&#39;][&#39;Protein##&#39; +
                                                          i] = struct.ProcessInfo.Elements[j][&#39;Composition&#39;].pop(i)

    for j in struct.EnzymeInfo.Elements.keys():
        for i in struct.EnzymeInfo.Elements[j][&#39;Subunits&#39;].keys():
            struct.EnzymeInfo.Elements[j][&#39;Subunits&#39;][&#39;Protein##&#39; +
                                                      i] = struct.EnzymeInfo.Elements[j][&#39;Subunits&#39;].pop(i)
        for i in range(len(struct.EnzymeInfo.Elements[j][&#39;Isozymes&#39;])):
            struct.EnzymeInfo.Elements[j][&#39;Isozymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.EnzymeInfo.Elements[j][&#39;Isozymes&#39;][i]
        for i in range(len(struct.EnzymeInfo.Elements[j][&#39;IdenticalEnzymes&#39;])):
            struct.EnzymeInfo.Elements[j][&#39;IdenticalEnzymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.EnzymeInfo.Elements[j][&#39;IdenticalEnzymes&#39;][i]
        for i in range(len(struct.EnzymeInfo.Elements[j][&#39;EnzymeCompartment&#39;])):
            struct.EnzymeInfo.Elements[j][&#39;EnzymeCompartment&#39;][i] = &#39;Compartment##&#39; + \
                struct.EnzymeInfo.Elements[j][&#39;EnzymeCompartment&#39;][i]
        struct.EnzymeInfo.Elements[j][&#39;Reaction&#39;] = &#39;Reaction##&#39; + \
            struct.EnzymeInfo.Elements[j][&#39;Reaction&#39;]

    for j in struct.ReactionInfo.Elements.keys():
        for i in struct.ReactionInfo.Elements[j][&#39;Reactants&#39;].keys():
            struct.ReactionInfo.Elements[j][&#39;Reactants&#39;][&#39;Metabolite##&#39; +
                                                         i] = struct.ReactionInfo.Elements[j][&#39;Reactants&#39;].pop(i)
        for i in struct.ReactionInfo.Elements[j][&#39;Products&#39;].keys():
            struct.ReactionInfo.Elements[j][&#39;Products&#39;][&#39;Metabolite##&#39; +
                                                        i] = struct.ReactionInfo.Elements[j][&#39;Products&#39;].pop(i)
        for i in range(len(struct.ReactionInfo.Elements[j][&#39;Twins&#39;])):
            struct.ReactionInfo.Elements[j][&#39;Twins&#39;][i] = &#39;Reaction##&#39; + \
                struct.ReactionInfo.Elements[j][&#39;Twins&#39;][i]
        struct.ReactionInfo.Elements[j][&#39;Enzyme&#39;] = &#39;Enzyme##&#39; + \
            struct.ReactionInfo.Elements[j][&#39;Enzyme&#39;]

    for j in struct.ProteinInfo.Elements.keys():
        for i in struct.ProteinInfo.Elements[j][&#39;ProcessRequirements&#39;].keys():
            struct.ProteinInfo.Elements[j][&#39;ProcessRequirements&#39;][&#39;Process##&#39; +
                                                                  i] = struct.ProteinInfo.Elements[j][&#39;ProcessRequirements&#39;].pop(i)
        for i in range(len(struct.ProteinInfo.Elements[j][&#39;associatedReactions&#39;])):
            struct.ProteinInfo.Elements[j][&#39;associatedReactions&#39;][i] = &#39;Reaction##&#39; + \
                struct.ProteinInfo.Elements[j][&#39;associatedReactions&#39;][i]
        for i in range(len(struct.ProteinInfo.Elements[j][&#39;associatedEnzymes&#39;])):
            struct.ProteinInfo.Elements[j][&#39;associatedEnzymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.ProteinInfo.Elements[j][&#39;associatedEnzymes&#39;][i]
        for i in range(len(struct.ProteinInfo.Elements[j][&#39;SupportsProcess&#39;])):
            struct.ProteinInfo.Elements[j][&#39;SupportsProcess&#39;][i] = &#39;Process##&#39; + \
                struct.ProteinInfo.Elements[j][&#39;SupportsProcess&#39;][i]
        struct.ProteinInfo.Elements[j][&#39;Compartment&#39;] = &#39;Compartment##&#39; + \
            struct.ProteinInfo.Elements[j][&#39;Compartment&#39;]

    for j in struct.CompartmentInfo.Elements.keys():
        for i in range(len(struct.CompartmentInfo.Elements[j][&#39;associatedProteins&#39;])):
            struct.CompartmentInfo.Elements[j][&#39;associatedProteins&#39;][i] = &#39;Protein##&#39; + \
                struct.CompartmentInfo.Elements[j][&#39;associatedProteins&#39;][i]
        for i in range(len(struct.CompartmentInfo.Elements[j][&#39;associatedReactions&#39;])):
            struct.CompartmentInfo.Elements[j][&#39;associatedReactions&#39;][i] = &#39;Reaction##&#39; + \
                struct.CompartmentInfo.Elements[j][&#39;associatedReactions&#39;][i]
        for i in range(len(struct.CompartmentInfo.Elements[j][&#39;associatedEnzymes&#39;])):
            struct.CompartmentInfo.Elements[j][&#39;associatedEnzymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.CompartmentInfo.Elements[j][&#39;associatedEnzymes&#39;][i]

    for j in struct.MetaboliteInfo.Elements.keys():
        for i in range(len(struct.MetaboliteInfo.Elements[j][&#39;ReactionsInvolvedWith&#39;])):
            struct.MetaboliteInfo.Elements[j][&#39;ReactionsInvolvedWith&#39;][i] = &#39;Reaction##&#39; + \
                struct.MetaboliteInfo.Elements[j][&#39;ReactionsInvolvedWith&#39;][i]

    for j in struct.MetaboliteConstraintsInfo.Elements.keys():
        struct.MetaboliteConstraintsInfo.Elements[j][&#39;AssociatedMetabolite&#39;] = &#39;Metabolite##&#39; + \
            struct.MetaboliteConstraintsInfo.Elements[j][&#39;AssociatedMetabolite&#39;]
    for j in struct.EnzymeConstraintsInfo.Elements.keys():
        struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedEnzyme&#39;] = &#39;Enzyme##&#39; + \
            struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedEnzyme&#39;]
    for j in struct.EnzymeConstraintsInfo.Elements.keys():
        struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedReaction&#39;] = &#39;Reaction##&#39; + \
            struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedReaction&#39;]
    for j in struct.ProcessConstraintsInfo.Elements.keys():
        struct.ProcessConstraintsInfo.Elements[j][&#39;AssociatedProcess&#39;] = &#39;Process##&#39; + \
            struct.ProcessConstraintsInfo.Elements[j][&#39;AssociatedProcess&#39;]
    for j in struct.DensityConstraintsInfo.Elements.keys():
        struct.DensityConstraintsInfo.Elements[j][&#39;AssociatedCompartment&#39;] = &#39;Compartment##&#39; + \
            struct.DensityConstraintsInfo.Elements[j][&#39;AssociatedCompartment&#39;]

    for i in struct.CompartmentInfo.Elements.keys():
        struct.CompartmentInfo.Elements[&#39;ID_&#39; + i] = struct.CompartmentInfo.Elements.pop(i)
    for i in struct.ProcessInfo.Elements.keys():
        struct.ProcessInfo.Elements[&#39;ID_&#39; + i] = struct.ProcessInfo.Elements.pop(i)
    for i in struct.MetaboliteInfo.Elements.keys():
        struct.MetaboliteInfo.Elements[&#39;ID_&#39; + i] = struct.MetaboliteInfo.Elements.pop(i)
    for i in struct.ModuleInfo.Elements.keys():
        struct.ModuleInfo.Elements[&#39;ID_&#39; + i] = struct.ModuleInfo.Elements.pop(i)
    for i in struct.EnzymeInfo.Elements.keys():
        struct.EnzymeInfo.Elements[&#39;ID_&#39; + i] = struct.EnzymeInfo.Elements.pop(i)
    for i in struct.ProteinInfo.Elements.keys():
        struct.ProteinInfo.Elements[&#39;ID_&#39; + i] = struct.ProteinInfo.Elements.pop(i)
    for i in struct.ReactionInfo.Elements.keys():
        struct.ReactionInfo.Elements[&#39;ID_&#39; + i] = struct.ReactionInfo.Elements.pop(i)
    for i in struct.MetaboliteConstraintsInfo.Elements.keys():
        struct.MetaboliteConstraintsInfo.Elements[&#39;ID_&#39; +
                                                  i] = struct.MetaboliteConstraintsInfo.Elements.pop(i)
    for i in struct.EnzymeConstraintsInfo.Elements.keys():
        struct.EnzymeConstraintsInfo.Elements[&#39;ID_&#39; +
                                              i] = struct.EnzymeConstraintsInfo.Elements.pop(i)
    for i in struct.ProcessConstraintsInfo.Elements.keys():
        struct.ProcessConstraintsInfo.Elements[&#39;ID_&#39; +
                                               i] = struct.ProcessConstraintsInfo.Elements.pop(i)
    for i in struct.DensityConstraintsInfo.Elements.keys():
        struct.DensityConstraintsInfo.Elements[&#39;ID_&#39; +
                                               i] = struct.DensityConstraintsInfo.Elements.pop(i)

    struct.ProcessInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.CompartmentInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.MetaboliteInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ModuleInfo.Elements.update({&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.EnzymeInfo.Elements.update({&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ProteinInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ReactionInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.DensityConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ProcessConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.MetaboliteConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.EnzymeConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})

    Block = {&#39;ModelInformation&#39;: struct.GeneralInfo.JSONize(),
             &#39;ModelStatistics&#39;: struct.ModelStatistics.JSONize(),
             &#39;Process&#39;: struct.ProcessInfo.JSONize(),
             &#39;Compartment&#39;: struct.CompartmentInfo.JSONize(),
             &#39;Metabolite&#39;: struct.MetaboliteInfo.JSONize(),
             &#39;Module&#39;: struct.ModuleInfo.JSONize(),
             &#39;Enzyme&#39;: struct.EnzymeInfo.JSONize(),
             &#39;Protein&#39;: struct.ProteinInfo.JSONize(),
             &#39;Reaction&#39;: struct.ReactionInfo.JSONize(),
             &#39;DensityConstraint&#39;: struct.DensityConstraintsInfo.JSONize(),
             &#39;ProcessConstraint&#39;: struct.ProcessConstraintsInfo.JSONize(),
             &#39;MetaboliteConstraint&#39;: struct.MetaboliteConstraintsInfo.JSONize(),
             &#39;EnzymeConstraint&#39;: struct.EnzymeConstraintsInfo.JSONize()
             }

    return({&#39;RBA_ModelData&#39;: {&#39;StructuralInformation&#39;: Block}})


def generateOverview(StatsReactions, StatsMetabolites, StatsModules, StatsEnzymes, StatsProteins, StatsMacromolecules, StatsProcesses, StatsCompartments, StatsTargets, StatsConstraintsBiological, StatsConstraintsMathematical):
    out = {&#39;Reactions Total&#39;: StatsReactions[&#39;ReactionsTotal&#39;],
           &#39;Reactions Unique&#39;: StatsReactions[&#39;ReactionsUnique&#39;],
           &#39;Reactions Spontaneous&#39;: StatsReactions[&#39;ReactionsSpontaneous&#39;],
           &#39;Reactions Enzymatic&#39;: StatsReactions[&#39;ReactionsEnzymatic&#39;],
           &#39;Reactions Internal&#39;: StatsReactions[&#39;ReactionsInternal&#39;],
           &#39;Reactions Exchange&#39;: StatsReactions[&#39;ReactionsExchange&#39;],
           &#39;Reactions CompartmentTransport&#39;: StatsReactions[&#39;ReactionsCompartmentTransport&#39;],
           &#39;Reactions Reversible&#39;: StatsReactions[&#39;ReactionsReversible&#39;],
           &#39;Reactions Irreversible&#39;: StatsReactions[&#39;ReactionsIrreversible&#39;],
           &#39;Metabolites Total&#39;: StatsMetabolites[&#39;MetabolitesTotal&#39;],
           &#39;Metabolites Internal&#39;: StatsMetabolites[&#39;MetabolitesInternal&#39;],
           &#39;Metabolites External&#39;: StatsMetabolites[&#39;MetabolitesExternal&#39;],
           &#39;Metabolites GrowthRelevant&#39;: StatsMetabolites[&#39;MetabolitesGrowthRelevant&#39;],
           &#39;Boundary Metabolites&#39;: StatsMetabolites[&#39;BoundaryMetabolites&#39;],
           &#39;Enzymes Total&#39;: StatsEnzymes[&#39;EnzymesTotal&#39;],
           &#39;Enzymes Unique&#39;: StatsEnzymes[&#39;EnzymesUnique&#39;],
           &#39;Proteins Total&#39;: StatsProteins[&#39;ProteinsTotal&#39;],
           &#39;RNAs Total&#39;: StatsMacromolecules[&#39;RNAsTotal&#39;],
           &#39;DNAs Total&#39;: StatsMacromolecules[&#39;DNAsTotal&#39;],
           &#39;Processes Total&#39;: StatsProcesses[&#39;ProcessesTotal&#39;],
           &#39;Modules Total&#39;: StatsModules[&#39;ModulesTotal&#39;],
           &#39;Compartments Total&#39;: StatsCompartments[&#39;CompartmentsTotal&#39;],
           &#39;Biological constraints metabolite&#39;: StatsConstraintsBiological[&#39;BioConstraintsMetabolite&#39;],
           &#39;Biological constraints capacity&#39;: StatsConstraintsBiological[&#39;BioConstraintsCapacity&#39;],
           &#39;Biological constraints process&#39;: StatsConstraintsBiological[&#39;BioConstraintsProcess&#39;],
           &#39;Biological constraints density&#39;: StatsConstraintsBiological[&#39;BioConstraintsDensity&#39;],
           &#39;Mathematical constraints variables&#39;: StatsConstraintsMathematical[&#39;MathConstraintsVariables&#39;],
           &#39;Mathematical constraints constraints&#39;: StatsConstraintsMathematical[&#39;MathConstraintsConstraints&#39;],
           &#39;Mathematical constraints equalities&#39;: StatsConstraintsMathematical[&#39;MathConstraintsEqualities&#39;],
           &#39;Mathematical constraints inequalities&#39;: StatsConstraintsMathematical[&#39;MathConstraintsInequalities&#39;]}
    out.update(StatsTargets)
    return(out)


def StatsConstraintsBiological(MetCs, CapCs, DenCs, EffCs):
    out = {&#39;BioConstraintsMetabolite&#39;: len(MetCs.keys()),
           &#39;BioConstraintsCapacity&#39;: len(CapCs.keys()),
           &#39;BioConstraintsProcess&#39;: len(EffCs.keys()),
           &#39;BioConstraintsDensity&#39;: len(DenCs.keys())}
    return(out)


def StatsConstraintsMathematical(MetCs, CapCs, DenCs, EffCs, Enzymes, Reactions, Processes):
    nVars = len(Enzymes.keys())+len(Reactions.keys())+len(Processes.keys())
    nConsts = len(MetCs.keys())+len(CapCs.keys())+len(DenCs.keys())+len(EffCs.keys())
    nEqC = 0
    nInC = 0
    for i in MetCs.keys():
        if MetCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if MetCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    for i in CapCs.keys():
        if CapCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if CapCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    for i in DenCs.keys():
        if DenCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if DenCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    for i in EffCs.keys():
        if EffCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if EffCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    out = {&#39;MathConstraintsVariables&#39;: nVars,
           &#39;MathConstraintsConstraints&#39;: nConsts,
           &#39;MathConstraintsEqualities&#39;: nEqC,
           &#39;MathConstraintsInequalities&#39;: nInC}
    return(out)


def findContainedProteins(comp, Prots):
    out = []
    for k in Prots.keys():
        if Prots[k][&#39;Compartment&#39;] == comp:
            out.append(k)
    return(out)


def findContainedMacromolecules(comp, Macromolecules):
    out = []
    for k in Macromolecules.keys():
        if Macromolecules[k][&#39;Compartment&#39;] == comp:
            out.append(k)
    return(out)


def findContainedEnzymes(ContainedProteins, Prots, Enzymes):
    rs = []
    enzs = []
    for k in ContainedProteins:
        rs = rs+Prots[k][&#39;associatedReactions&#39;]
        enzs = enzs+Prots[k][&#39;associatedEnzymes&#39;]
    out = {&#39;containedEnzymes&#39;: list(numpy.unique(enzs)),
           &#39;containedReactions&#39;: list(numpy.unique(rs))}
    return(out)


def findAssociatedEnzyme(Enzymes, protein):
    out1 = []
    out2 = []
    for e in Enzymes.keys():
        if protein in Enzymes[e][&#39;Subunits&#39;].keys():
            out1.append(e)
            out2.append(Enzymes[e][&#39;Reaction&#39;])
    out = {&#39;Enz&#39;: out1,
           &#39;Rx&#39;: out2}
    return(out)


def findIsozymes(ez, blocks, Reactions, rx):
    out = []
    twins = Reactions[rx][&#39;Twins&#39;]
    if len(twins) &gt; 0:
        for r in twins:
            if not type(r) == list:
                if not type(Reactions[r][&#39;Enzyme&#39;]) == list:
                    out.append(Reactions[r][&#39;Enzyme&#39;])
    return(out)


def sortConstraints(matrix, model):
    metaboliteList = [model.metabolism.species._elements[m].id for m in range(
        len(model.metabolism.species._elements))]
    mets = {}
    capacity = {}
    efficiency = {}
    density = {}
    for j in range(len(matrix.row_names)):
        i = matrix.row_names[j]
        if i in metaboliteList:
            mets[i] = j
        if &#39;enzyme&#39; in i:
            capacity[i] = j
        if i.startswith(&#39;P_&#39;):
            efficiency[i] = j
        if &#39;_density&#39; in i:
            density[i] = j
    out = {&#39;MetaboliteConsts&#39;: mets,
           &#39;ProcessConsts&#39;: efficiency,
           &#39;EnzymeConsts&#39;: capacity,
           &#39;DensityConsts&#39;: density}
    return(out)


def JSON_Int64_compensation(o):
    if isinstance(o, numpy.int64):
        return int(o)
#    raise TypeError</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="rbatools.rba_ModelStructure.JSON_Int64_compensation"><code class="name flex">
<span>def <span class="ident">JSON_Int64_compensation</span></span>(<span>o)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def JSON_Int64_compensation(o):
    if isinstance(o, numpy.int64):
        return int(o)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.StatsConstraintsBiological"><code class="name flex">
<span>def <span class="ident">StatsConstraintsBiological</span></span>(<span>MetCs, CapCs, DenCs, EffCs)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def StatsConstraintsBiological(MetCs, CapCs, DenCs, EffCs):
    out = {&#39;BioConstraintsMetabolite&#39;: len(MetCs.keys()),
           &#39;BioConstraintsCapacity&#39;: len(CapCs.keys()),
           &#39;BioConstraintsProcess&#39;: len(EffCs.keys()),
           &#39;BioConstraintsDensity&#39;: len(DenCs.keys())}
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.StatsConstraintsMathematical"><code class="name flex">
<span>def <span class="ident">StatsConstraintsMathematical</span></span>(<span>MetCs, CapCs, DenCs, EffCs, Enzymes, Reactions, Processes)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def StatsConstraintsMathematical(MetCs, CapCs, DenCs, EffCs, Enzymes, Reactions, Processes):
    nVars = len(Enzymes.keys())+len(Reactions.keys())+len(Processes.keys())
    nConsts = len(MetCs.keys())+len(CapCs.keys())+len(DenCs.keys())+len(EffCs.keys())
    nEqC = 0
    nInC = 0
    for i in MetCs.keys():
        if MetCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if MetCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    for i in CapCs.keys():
        if CapCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if CapCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    for i in DenCs.keys():
        if DenCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if DenCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    for i in EffCs.keys():
        if EffCs[i][&#39;Type&#39;] == &#39;&lt;=&#39;:
            nInC += 1
        if EffCs[i][&#39;Type&#39;] == &#39;=&#39;:
            nEqC += 1
    out = {&#39;MathConstraintsVariables&#39;: nVars,
           &#39;MathConstraintsConstraints&#39;: nConsts,
           &#39;MathConstraintsEqualities&#39;: nEqC,
           &#39;MathConstraintsInequalities&#39;: nInC}
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.findAssociatedEnzyme"><code class="name flex">
<span>def <span class="ident">findAssociatedEnzyme</span></span>(<span>Enzymes, protein)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def findAssociatedEnzyme(Enzymes, protein):
    out1 = []
    out2 = []
    for e in Enzymes.keys():
        if protein in Enzymes[e][&#39;Subunits&#39;].keys():
            out1.append(e)
            out2.append(Enzymes[e][&#39;Reaction&#39;])
    out = {&#39;Enz&#39;: out1,
           &#39;Rx&#39;: out2}
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.findContainedEnzymes"><code class="name flex">
<span>def <span class="ident">findContainedEnzymes</span></span>(<span>ContainedProteins, Prots, Enzymes)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def findContainedEnzymes(ContainedProteins, Prots, Enzymes):
    rs = []
    enzs = []
    for k in ContainedProteins:
        rs = rs+Prots[k][&#39;associatedReactions&#39;]
        enzs = enzs+Prots[k][&#39;associatedEnzymes&#39;]
    out = {&#39;containedEnzymes&#39;: list(numpy.unique(enzs)),
           &#39;containedReactions&#39;: list(numpy.unique(rs))}
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.findContainedMacromolecules"><code class="name flex">
<span>def <span class="ident">findContainedMacromolecules</span></span>(<span>comp, Macromolecules)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def findContainedMacromolecules(comp, Macromolecules):
    out = []
    for k in Macromolecules.keys():
        if Macromolecules[k][&#39;Compartment&#39;] == comp:
            out.append(k)
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.findContainedProteins"><code class="name flex">
<span>def <span class="ident">findContainedProteins</span></span>(<span>comp, Prots)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def findContainedProteins(comp, Prots):
    out = []
    for k in Prots.keys():
        if Prots[k][&#39;Compartment&#39;] == comp:
            out.append(k)
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.findIsozymes"><code class="name flex">
<span>def <span class="ident">findIsozymes</span></span>(<span>ez, blocks, Reactions, rx)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def findIsozymes(ez, blocks, Reactions, rx):
    out = []
    twins = Reactions[rx][&#39;Twins&#39;]
    if len(twins) &gt; 0:
        for r in twins:
            if not type(r) == list:
                if not type(Reactions[r][&#39;Enzyme&#39;]) == list:
                    out.append(Reactions[r][&#39;Enzyme&#39;])
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.findParameterDependencies"><code class="name flex">
<span>def <span class="ident">findParameterDependencies</span></span>(<span>ModelStructure)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def findParameterDependencies(ModelStructure):
    MedDepts = {}
    MuDepts = []
    for eK in ModelStructure.EnzymeConstraintsInfo.Elements.keys():
        e = ModelStructure.EnzymeConstraintsInfo.Elements[eK]
        for pf in e[&#39;CapacityParameter&#39;]:
            iV = list(pf.values())[0][&#39;IndependentVariable&#39;]
            if list(pf.values())[0][&#39;FunctionType&#39;] != &#39;constant&#39;:
                if iV.startswith(&#39;M_&#39;):
                    if iV.rsplit(&#39;_&#39;, 1)[0] in MedDepts.keys():
                        MedDepts[iV.rsplit(&#39;_&#39;, 1)[0]].append(e[&#39;ID&#39;])
                    elif iV.rsplit(&#39;_&#39;, 1)[0] not in MedDepts.keys():
                        MedDepts.update({iV.rsplit(&#39;_&#39;, 1)[0]: [e[&#39;ID&#39;]]})
                if iV == &#39;growth_rate&#39;:
                    if e[&#39;ID&#39;] not in MuDepts:
                        MuDepts.append(e[&#39;ID&#39;])
    for dK in ModelStructure.DensityConstraintsInfo.Elements.keys():
        d = ModelStructure.DensityConstraintsInfo.Elements[dK]
        for pf in d[&#39;CapacityParameter&#39;]:
            iV = list(pf.values())[0][&#39;IndependentVariable&#39;]
            if list(pf.values())[0][&#39;FunctionType&#39;] != &#39;constant&#39;:
                if iV.startswith(&#39;M_&#39;):
                    if iV.rsplit(&#39;_&#39;, 1)[0] in MedDepts.keys():
                        MedDepts[iV.rsplit(&#39;_&#39;, 1)[0]].append(d[&#39;ID&#39;])
                    elif iV.rsplit(&#39;_&#39;, 1)[0] not in MedDepts.keys():
                        MedDepts.update({iV.rsplit(&#39;_&#39;, 1)[0]: [d[&#39;ID&#39;]]})
                if iV == &#39;growth_rate&#39;:
                    if d[&#39;ID&#39;] not in MuDepts:
                        MuDepts.append(d[&#39;ID&#39;])
    for pK in ModelStructure.ProcessConstraintsInfo.Elements.keys():
        p = ModelStructure.ProcessConstraintsInfo.Elements[pK]
        for pf in p[&#39;CapacityParameter&#39;]:
            iV = list(pf.values())[0][&#39;IndependentVariable&#39;]
            if list(pf.values())[0][&#39;FunctionType&#39;] != &#39;constant&#39;:
                if iV.startswith(&#39;M_&#39;):
                    if iV.rsplit(&#39;_&#39;, 1)[0] in MedDepts.keys():
                        MedDepts[iV.rsplit(&#39;_&#39;, 1)[0]].append(p[&#39;ID&#39;])
                    elif iV.rsplit(&#39;_&#39;, 1)[0] not in MedDepts.keys():
                        MedDepts.update({iV.rsplit(&#39;_&#39;, 1)[0]: [p[&#39;ID&#39;]]})
                if iV == &#39;growth_rate&#39;:
                    if p[&#39;ID&#39;] not in MuDepts:
                        MuDepts.append(p[&#39;ID&#39;])
    return(MedDepts, MuDepts)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.generateOverview"><code class="name flex">
<span>def <span class="ident">generateOverview</span></span>(<span>StatsReactions, StatsMetabolites, StatsModules, StatsEnzymes, StatsProteins, StatsMacromolecules, StatsProcesses, StatsCompartments, StatsTargets, StatsConstraintsBiological, StatsConstraintsMathematical)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def generateOverview(StatsReactions, StatsMetabolites, StatsModules, StatsEnzymes, StatsProteins, StatsMacromolecules, StatsProcesses, StatsCompartments, StatsTargets, StatsConstraintsBiological, StatsConstraintsMathematical):
    out = {&#39;Reactions Total&#39;: StatsReactions[&#39;ReactionsTotal&#39;],
           &#39;Reactions Unique&#39;: StatsReactions[&#39;ReactionsUnique&#39;],
           &#39;Reactions Spontaneous&#39;: StatsReactions[&#39;ReactionsSpontaneous&#39;],
           &#39;Reactions Enzymatic&#39;: StatsReactions[&#39;ReactionsEnzymatic&#39;],
           &#39;Reactions Internal&#39;: StatsReactions[&#39;ReactionsInternal&#39;],
           &#39;Reactions Exchange&#39;: StatsReactions[&#39;ReactionsExchange&#39;],
           &#39;Reactions CompartmentTransport&#39;: StatsReactions[&#39;ReactionsCompartmentTransport&#39;],
           &#39;Reactions Reversible&#39;: StatsReactions[&#39;ReactionsReversible&#39;],
           &#39;Reactions Irreversible&#39;: StatsReactions[&#39;ReactionsIrreversible&#39;],
           &#39;Metabolites Total&#39;: StatsMetabolites[&#39;MetabolitesTotal&#39;],
           &#39;Metabolites Internal&#39;: StatsMetabolites[&#39;MetabolitesInternal&#39;],
           &#39;Metabolites External&#39;: StatsMetabolites[&#39;MetabolitesExternal&#39;],
           &#39;Metabolites GrowthRelevant&#39;: StatsMetabolites[&#39;MetabolitesGrowthRelevant&#39;],
           &#39;Boundary Metabolites&#39;: StatsMetabolites[&#39;BoundaryMetabolites&#39;],
           &#39;Enzymes Total&#39;: StatsEnzymes[&#39;EnzymesTotal&#39;],
           &#39;Enzymes Unique&#39;: StatsEnzymes[&#39;EnzymesUnique&#39;],
           &#39;Proteins Total&#39;: StatsProteins[&#39;ProteinsTotal&#39;],
           &#39;RNAs Total&#39;: StatsMacromolecules[&#39;RNAsTotal&#39;],
           &#39;DNAs Total&#39;: StatsMacromolecules[&#39;DNAsTotal&#39;],
           &#39;Processes Total&#39;: StatsProcesses[&#39;ProcessesTotal&#39;],
           &#39;Modules Total&#39;: StatsModules[&#39;ModulesTotal&#39;],
           &#39;Compartments Total&#39;: StatsCompartments[&#39;CompartmentsTotal&#39;],
           &#39;Biological constraints metabolite&#39;: StatsConstraintsBiological[&#39;BioConstraintsMetabolite&#39;],
           &#39;Biological constraints capacity&#39;: StatsConstraintsBiological[&#39;BioConstraintsCapacity&#39;],
           &#39;Biological constraints process&#39;: StatsConstraintsBiological[&#39;BioConstraintsProcess&#39;],
           &#39;Biological constraints density&#39;: StatsConstraintsBiological[&#39;BioConstraintsDensity&#39;],
           &#39;Mathematical constraints variables&#39;: StatsConstraintsMathematical[&#39;MathConstraintsVariables&#39;],
           &#39;Mathematical constraints constraints&#39;: StatsConstraintsMathematical[&#39;MathConstraintsConstraints&#39;],
           &#39;Mathematical constraints equalities&#39;: StatsConstraintsMathematical[&#39;MathConstraintsEqualities&#39;],
           &#39;Mathematical constraints inequalities&#39;: StatsConstraintsMathematical[&#39;MathConstraintsInequalities&#39;]}
    out.update(StatsTargets)
    return(out)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.generateProteinGeneMatrix"><code class="name flex">
<span>def <span class="ident">generateProteinGeneMatrix</span></span>(<span>ModelStructure)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def generateProteinGeneMatrix(ModelStructure):
    uniqueProteins = []
    uniqueProteinMap = {}
    Proteins = list(ModelStructure.ProteinInfo.Elements.keys())
    for i in ModelStructure.ProteinInfo.Elements.keys():
        if ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;] not in list(uniqueProteinMap.keys()):
            uniqueProteinMap.update({ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;]: []})
            uniqueProteins.append(ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;])
        uniqueProteinMap[ModelStructure.ProteinInfo.Elements[i][&#39;ProtoID&#39;]].append(i)
    ProteinProteinMatrix = numpy.zeros(
        (len(list(uniqueProteinMap.keys())), len(list(ModelStructure.ProteinInfo.Elements.keys()))))
    for u in list(uniqueProteinMap.keys()):
        row_ind = uniqueProteins.index(u)
        for i in uniqueProteinMap[u]:
            col_ind = Proteins.index(i)
            ProteinProteinMatrix[row_ind, col_ind] = 1
    return({&#39;Matrix&#39;: numpy.array(ProteinProteinMatrix), &#39;Proteins&#39;: Proteins, &#39;ProtoProteins&#39;: uniqueProteins})</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.generateProteinMatrix"><code class="name flex">
<span>def <span class="ident">generateProteinMatrix</span></span>(<span>ModelStructure)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def generateProteinMatrix(ModelStructure):
    Proteins = list(ModelStructure.ProteinInfo.Elements.keys())
    # print(list(ModelStructure.ProcessInfo.Elements.keys()))
    Processes = [ModelStructure.ProcessInfo.Elements[i][&#39;ID&#39;] +
                 &#39;_machinery&#39; for i in list(ModelStructure.ProcessInfo.Elements.keys())]
    Enzymes = list(ModelStructure.EnzymeInfo.Elements.keys())
    Consumers = list(set(list(Enzymes+Processes)))
    ProteinMatrix = numpy.zeros((len(Proteins), len(Consumers)))
    for p in Proteins:
        if len(ModelStructure.ProteinInfo.Elements[p][&#39;SupportsProcess&#39;]) &gt; 0:
            # print(list(ModelStructure.ProteinInfo.Elements[p][&#39;SupportsProcess&#39;]))
            for pc in list(ModelStructure.ProteinInfo.Elements[p][&#39;SupportsProcess&#39;]):
                coeff = 0
                row_ind = Proteins.index(p)
                col_ind = Consumers.index(
                    ModelStructure.ProcessInfo.Elements[pc][&#39;ID&#39;]+&#39;_machinery&#39;)
                coeff = ModelStructure.ProcessInfo.Elements[pc][&#39;Composition&#39;][p]
                ProteinMatrix[row_ind, col_ind] += coeff
        if len(ModelStructure.ProteinInfo.Elements[p][&#39;associatedEnzymes&#39;]) &gt; 0:
            for ez in list(ModelStructure.ProteinInfo.Elements[p][&#39;associatedEnzymes&#39;]):
                coeff = 0
                row_ind = Proteins.index(p)
                col_ind = Consumers.index(ez)
                coeff = ModelStructure.EnzymeInfo.Elements[ez][&#39;Subunits&#39;][p]
                ProteinMatrix[row_ind, col_ind] += coeff
    return({&#39;Matrix&#39;: numpy.array(ProteinMatrix), &#39;Consumers&#39;: Consumers, &#39;Proteins&#39;: Proteins})</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.htmlStyle"><code class="name flex">
<span>def <span class="ident">htmlStyle</span></span>(<span>structOriginal)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def htmlStyle(structOriginal):
    struct = copy.deepcopy(structOriginal)

    for j in struct.ProcessInfo.Elements.keys():
        for i in struct.ProcessInfo.Elements[j][&#39;Composition&#39;].keys():
            struct.ProcessInfo.Elements[j][&#39;Composition&#39;][&#39;Protein##&#39; +
                                                          i] = struct.ProcessInfo.Elements[j][&#39;Composition&#39;].pop(i)

    for j in struct.EnzymeInfo.Elements.keys():
        for i in struct.EnzymeInfo.Elements[j][&#39;Subunits&#39;].keys():
            struct.EnzymeInfo.Elements[j][&#39;Subunits&#39;][&#39;Protein##&#39; +
                                                      i] = struct.EnzymeInfo.Elements[j][&#39;Subunits&#39;].pop(i)
        for i in range(len(struct.EnzymeInfo.Elements[j][&#39;Isozymes&#39;])):
            struct.EnzymeInfo.Elements[j][&#39;Isozymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.EnzymeInfo.Elements[j][&#39;Isozymes&#39;][i]
        for i in range(len(struct.EnzymeInfo.Elements[j][&#39;IdenticalEnzymes&#39;])):
            struct.EnzymeInfo.Elements[j][&#39;IdenticalEnzymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.EnzymeInfo.Elements[j][&#39;IdenticalEnzymes&#39;][i]
        for i in range(len(struct.EnzymeInfo.Elements[j][&#39;EnzymeCompartment&#39;])):
            struct.EnzymeInfo.Elements[j][&#39;EnzymeCompartment&#39;][i] = &#39;Compartment##&#39; + \
                struct.EnzymeInfo.Elements[j][&#39;EnzymeCompartment&#39;][i]
        struct.EnzymeInfo.Elements[j][&#39;Reaction&#39;] = &#39;Reaction##&#39; + \
            struct.EnzymeInfo.Elements[j][&#39;Reaction&#39;]

    for j in struct.ReactionInfo.Elements.keys():
        for i in struct.ReactionInfo.Elements[j][&#39;Reactants&#39;].keys():
            struct.ReactionInfo.Elements[j][&#39;Reactants&#39;][&#39;Metabolite##&#39; +
                                                         i] = struct.ReactionInfo.Elements[j][&#39;Reactants&#39;].pop(i)
        for i in struct.ReactionInfo.Elements[j][&#39;Products&#39;].keys():
            struct.ReactionInfo.Elements[j][&#39;Products&#39;][&#39;Metabolite##&#39; +
                                                        i] = struct.ReactionInfo.Elements[j][&#39;Products&#39;].pop(i)
        for i in range(len(struct.ReactionInfo.Elements[j][&#39;Twins&#39;])):
            struct.ReactionInfo.Elements[j][&#39;Twins&#39;][i] = &#39;Reaction##&#39; + \
                struct.ReactionInfo.Elements[j][&#39;Twins&#39;][i]
        struct.ReactionInfo.Elements[j][&#39;Enzyme&#39;] = &#39;Enzyme##&#39; + \
            struct.ReactionInfo.Elements[j][&#39;Enzyme&#39;]

    for j in struct.ProteinInfo.Elements.keys():
        for i in struct.ProteinInfo.Elements[j][&#39;ProcessRequirements&#39;].keys():
            struct.ProteinInfo.Elements[j][&#39;ProcessRequirements&#39;][&#39;Process##&#39; +
                                                                  i] = struct.ProteinInfo.Elements[j][&#39;ProcessRequirements&#39;].pop(i)
        for i in range(len(struct.ProteinInfo.Elements[j][&#39;associatedReactions&#39;])):
            struct.ProteinInfo.Elements[j][&#39;associatedReactions&#39;][i] = &#39;Reaction##&#39; + \
                struct.ProteinInfo.Elements[j][&#39;associatedReactions&#39;][i]
        for i in range(len(struct.ProteinInfo.Elements[j][&#39;associatedEnzymes&#39;])):
            struct.ProteinInfo.Elements[j][&#39;associatedEnzymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.ProteinInfo.Elements[j][&#39;associatedEnzymes&#39;][i]
        for i in range(len(struct.ProteinInfo.Elements[j][&#39;SupportsProcess&#39;])):
            struct.ProteinInfo.Elements[j][&#39;SupportsProcess&#39;][i] = &#39;Process##&#39; + \
                struct.ProteinInfo.Elements[j][&#39;SupportsProcess&#39;][i]
        struct.ProteinInfo.Elements[j][&#39;Compartment&#39;] = &#39;Compartment##&#39; + \
            struct.ProteinInfo.Elements[j][&#39;Compartment&#39;]

    for j in struct.CompartmentInfo.Elements.keys():
        for i in range(len(struct.CompartmentInfo.Elements[j][&#39;associatedProteins&#39;])):
            struct.CompartmentInfo.Elements[j][&#39;associatedProteins&#39;][i] = &#39;Protein##&#39; + \
                struct.CompartmentInfo.Elements[j][&#39;associatedProteins&#39;][i]
        for i in range(len(struct.CompartmentInfo.Elements[j][&#39;associatedReactions&#39;])):
            struct.CompartmentInfo.Elements[j][&#39;associatedReactions&#39;][i] = &#39;Reaction##&#39; + \
                struct.CompartmentInfo.Elements[j][&#39;associatedReactions&#39;][i]
        for i in range(len(struct.CompartmentInfo.Elements[j][&#39;associatedEnzymes&#39;])):
            struct.CompartmentInfo.Elements[j][&#39;associatedEnzymes&#39;][i] = &#39;Enzyme##&#39; + \
                struct.CompartmentInfo.Elements[j][&#39;associatedEnzymes&#39;][i]

    for j in struct.MetaboliteInfo.Elements.keys():
        for i in range(len(struct.MetaboliteInfo.Elements[j][&#39;ReactionsInvolvedWith&#39;])):
            struct.MetaboliteInfo.Elements[j][&#39;ReactionsInvolvedWith&#39;][i] = &#39;Reaction##&#39; + \
                struct.MetaboliteInfo.Elements[j][&#39;ReactionsInvolvedWith&#39;][i]

    for j in struct.MetaboliteConstraintsInfo.Elements.keys():
        struct.MetaboliteConstraintsInfo.Elements[j][&#39;AssociatedMetabolite&#39;] = &#39;Metabolite##&#39; + \
            struct.MetaboliteConstraintsInfo.Elements[j][&#39;AssociatedMetabolite&#39;]
    for j in struct.EnzymeConstraintsInfo.Elements.keys():
        struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedEnzyme&#39;] = &#39;Enzyme##&#39; + \
            struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedEnzyme&#39;]
    for j in struct.EnzymeConstraintsInfo.Elements.keys():
        struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedReaction&#39;] = &#39;Reaction##&#39; + \
            struct.EnzymeConstraintsInfo.Elements[j][&#39;AssociatedReaction&#39;]
    for j in struct.ProcessConstraintsInfo.Elements.keys():
        struct.ProcessConstraintsInfo.Elements[j][&#39;AssociatedProcess&#39;] = &#39;Process##&#39; + \
            struct.ProcessConstraintsInfo.Elements[j][&#39;AssociatedProcess&#39;]
    for j in struct.DensityConstraintsInfo.Elements.keys():
        struct.DensityConstraintsInfo.Elements[j][&#39;AssociatedCompartment&#39;] = &#39;Compartment##&#39; + \
            struct.DensityConstraintsInfo.Elements[j][&#39;AssociatedCompartment&#39;]

    for i in struct.CompartmentInfo.Elements.keys():
        struct.CompartmentInfo.Elements[&#39;ID_&#39; + i] = struct.CompartmentInfo.Elements.pop(i)
    for i in struct.ProcessInfo.Elements.keys():
        struct.ProcessInfo.Elements[&#39;ID_&#39; + i] = struct.ProcessInfo.Elements.pop(i)
    for i in struct.MetaboliteInfo.Elements.keys():
        struct.MetaboliteInfo.Elements[&#39;ID_&#39; + i] = struct.MetaboliteInfo.Elements.pop(i)
    for i in struct.ModuleInfo.Elements.keys():
        struct.ModuleInfo.Elements[&#39;ID_&#39; + i] = struct.ModuleInfo.Elements.pop(i)
    for i in struct.EnzymeInfo.Elements.keys():
        struct.EnzymeInfo.Elements[&#39;ID_&#39; + i] = struct.EnzymeInfo.Elements.pop(i)
    for i in struct.ProteinInfo.Elements.keys():
        struct.ProteinInfo.Elements[&#39;ID_&#39; + i] = struct.ProteinInfo.Elements.pop(i)
    for i in struct.ReactionInfo.Elements.keys():
        struct.ReactionInfo.Elements[&#39;ID_&#39; + i] = struct.ReactionInfo.Elements.pop(i)
    for i in struct.MetaboliteConstraintsInfo.Elements.keys():
        struct.MetaboliteConstraintsInfo.Elements[&#39;ID_&#39; +
                                                  i] = struct.MetaboliteConstraintsInfo.Elements.pop(i)
    for i in struct.EnzymeConstraintsInfo.Elements.keys():
        struct.EnzymeConstraintsInfo.Elements[&#39;ID_&#39; +
                                              i] = struct.EnzymeConstraintsInfo.Elements.pop(i)
    for i in struct.ProcessConstraintsInfo.Elements.keys():
        struct.ProcessConstraintsInfo.Elements[&#39;ID_&#39; +
                                               i] = struct.ProcessConstraintsInfo.Elements.pop(i)
    for i in struct.DensityConstraintsInfo.Elements.keys():
        struct.DensityConstraintsInfo.Elements[&#39;ID_&#39; +
                                               i] = struct.DensityConstraintsInfo.Elements.pop(i)

    struct.ProcessInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.CompartmentInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.MetaboliteInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ModuleInfo.Elements.update({&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.EnzymeInfo.Elements.update({&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ProteinInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ReactionInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.DensityConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.ProcessConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.MetaboliteConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})
    struct.EnzymeConstraintsInfo.Elements.update(
        {&#39;Description&#39;: &#39;abcdefghijklmnopqrstuvwxyz&#39;, &#39;Pictures&#39;: []})

    Block = {&#39;ModelInformation&#39;: struct.GeneralInfo.JSONize(),
             &#39;ModelStatistics&#39;: struct.ModelStatistics.JSONize(),
             &#39;Process&#39;: struct.ProcessInfo.JSONize(),
             &#39;Compartment&#39;: struct.CompartmentInfo.JSONize(),
             &#39;Metabolite&#39;: struct.MetaboliteInfo.JSONize(),
             &#39;Module&#39;: struct.ModuleInfo.JSONize(),
             &#39;Enzyme&#39;: struct.EnzymeInfo.JSONize(),
             &#39;Protein&#39;: struct.ProteinInfo.JSONize(),
             &#39;Reaction&#39;: struct.ReactionInfo.JSONize(),
             &#39;DensityConstraint&#39;: struct.DensityConstraintsInfo.JSONize(),
             &#39;ProcessConstraint&#39;: struct.ProcessConstraintsInfo.JSONize(),
             &#39;MetaboliteConstraint&#39;: struct.MetaboliteConstraintsInfo.JSONize(),
             &#39;EnzymeConstraint&#39;: struct.EnzymeConstraintsInfo.JSONize()
             }

    return({&#39;RBA_ModelData&#39;: {&#39;StructuralInformation&#39;: Block}})</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importBiggMetabolites"><code class="name flex">
<span>def <span class="ident">importBiggMetabolites</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importBiggMetabolites(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/bigg_models_metabolites.txt&#39;)):
        return(pandas.read_csv(str(xml_dir+&#39;/bigg_models_metabolites.txt&#39;), sep=&#39;\t&#39;, index_col=0))
    else:
        sys.exit(&#39;\n Required BiGG Metabolite File &#34;bigg_models_metabolites.txt&#34; not found.\n&#39; +
                 &#39; Please provide in input-directory\n&#39; +
                 &#39; To be found under: http://bigg.ucsd.edu/static/namespace/bigg_models_metabolites.txt\n&#39;)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importBiggReactions"><code class="name flex">
<span>def <span class="ident">importBiggReactions</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importBiggReactions(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/bigg_models_reactions.txt&#39;)):
        return(pandas.read_csv(str(xml_dir+&#39;/bigg_models_reactions.txt&#39;), sep=&#39;\t&#39;, index_col=0))
    else:
        sys.exit(&#39;\n Required BiGG Reaction File &#34;bigg_models_reactions.txt &#34;not found.\n&#39; +
                 &#39; Please provide in input-directory\n&#39; +
                 &#39; To be found under: http://bigg.ucsd.edu/static/namespace/bigg_models_reactions.txt\n&#39;)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importGeneAnnotations"><code class="name flex">
<span>def <span class="ident">importGeneAnnotations</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importGeneAnnotations(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/GeneAnnotations.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/GeneAnnotations.csv&#39;), sep=&#39;,&#39;, index_col=0)
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;GeneAnnotations.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No Gene-annotation file &#34;GeneAnnotations.csv&#34; provided.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importMetaboliteAnnotations"><code class="name flex">
<span>def <span class="ident">importMetaboliteAnnotations</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importMetaboliteAnnotations(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/MetaboliteAnnotations.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/MetaboliteAnnotations.csv&#39;), sep=&#39;,&#39;, index_col=0)
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;MetaboliteAnnotations.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No Reaction-annotation file &#34;MetaboliteAnnotations.csv&#34; provided.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importModelInfo"><code class="name flex">
<span>def <span class="ident">importModelInfo</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importModelInfo(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/ModelInformation.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/ModelInformation.csv&#39;),
                              sep=&#39;,&#39;, header=0)
        out.index = list(out[&#39;Key&#39;])
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;ModelInformation.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No model-info file &#34;ModelInformation.csv&#34; provided.\n&#39; +
              &#39; Using dummy-information\n&#39;)
        return(pandas.DataFrame([[&#39;Name&#39;, &#39;ModelName&#39;], [&#39;Author&#39;, &#39;John Doe&#39;], [&#39;Organism&#39;, &#39;Life&#39;], [&#39;Reconstruction&#39;, &#39;GSMM&#39;], [&#39;SBML-file&#39;, &#39;Not Provided&#39;]], index=[&#39;Name&#39;, &#39;Author&#39;, &#39;Organism&#39;, &#39;Reconstruction&#39;, &#39;SBML-file&#39;], columns=[&#39;Key&#39;, &#39;Value&#39;]))</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importReactionAnnotations"><code class="name flex">
<span>def <span class="ident">importReactionAnnotations</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importReactionAnnotations(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/ReactionAnnotations.csv&#39;)):
        out = pandas.read_csv(str(xml_dir+&#39;/ReactionAnnotations.csv&#39;), sep=&#39;,&#39;, index_col=0)
        if len(list(out)) == 0:
            print(&#39;WARNING: Your file &#34;ReactionAnnotations.csv&#34; seems to be empty or has the wrong delimiter (comma required).&#39;)
        return(out)
    else:
        print(&#39;\n No Reaction-annotation file &#34;ReactionAnnotations.csv&#34; provided.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importSbmlFile"><code class="name flex">
<span>def <span class="ident">importSbmlFile</span></span>(<span>xml_dir, filename)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importSbmlFile(xml_dir, filename):
    if os.path.isfile(str(xml_dir+&#39;/&#39;+filename)):
        SBfile = libsbml.readSBML(str(xml_dir+&#39;/&#39;+filename))
        if SBfile.getNumErrors() &gt; 0:
            SBfile.printErrors()
            print(&#39;Invalid SBML&#39;)
            return(str(&#39;Not There&#39;))
        else:
            sbml = SBfile
            return(sbml)
    else:
        print(&#39;\n SBML-file {} not found.\n&#39; +
              &#39; Continuing without additional information...\n&#39;.format(filename))
        return(str(&#39;Not There&#39;))</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.importUniprotFile"><code class="name flex">
<span>def <span class="ident">importUniprotFile</span></span>(<span>xml_dir)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def importUniprotFile(xml_dir):
    if os.path.isfile(str(xml_dir+&#39;/uniprot.csv&#39;)):
        return(pandas.read_csv(str(xml_dir+&#39;/uniprot.csv&#39;), sep=&#39;\t&#39;))
    else:
        print(&#39;\n Uniprot-file &#34;uniprot.csv&#34; not found.\n&#39; +
              &#39; Continuing without additional information...\n&#39;)
        return(str(&#39;Not There&#39;))</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.sortConstraints"><code class="name flex">
<span>def <span class="ident">sortConstraints</span></span>(<span>matrix, model)</span>
</code></dt>
<dd>
<section class="desc"></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def sortConstraints(matrix, model):
    metaboliteList = [model.metabolism.species._elements[m].id for m in range(
        len(model.metabolism.species._elements))]
    mets = {}
    capacity = {}
    efficiency = {}
    density = {}
    for j in range(len(matrix.row_names)):
        i = matrix.row_names[j]
        if i in metaboliteList:
            mets[i] = j
        if &#39;enzyme&#39; in i:
            capacity[i] = j
        if i.startswith(&#39;P_&#39;):
            efficiency[i] = j
        if &#39;_density&#39; in i:
            density[i] = j
    out = {&#39;MetaboliteConsts&#39;: mets,
           &#39;ProcessConsts&#39;: efficiency,
           &#39;EnzymeConsts&#39;: capacity,
           &#39;DensityConsts&#39;: density}
    return(out)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="rbatools.rba_ModelStructure.RBA_ModelStructure"><code class="flex name class">
<span>class <span class="ident">RBA_ModelStructure</span></span>
</code></dt>
<dd>
<section class="desc"><p>Class holding information on model-structure.</p>
<h2 id="attributes">Attributes</h2>
<dl>
<dt><strong><code>GeneralInfo</code></strong> :&ensp;<code>rbatools.description_block.Description_block</code></dt>
<dd>Model description</dd>
<dt><strong><code>MetaboliteInfo</code></strong> :&ensp;<code>rbatools.metabolite_block.Metabolite_block</code></dt>
<dd>Metabolite information</dd>
<dt><strong><code>ModuleInfo</code></strong> :&ensp;<code>rbatools.module_block.Module_block</code></dt>
<dd>Module information</dd>
<dt><strong><code>ProcessInfo</code></strong> :&ensp;<code>rbatools.process_block.Process_block</code></dt>
<dd>Process information</dd>
<dt><strong><code>ReactionInfo</code></strong> :&ensp;<code>rbatools.reaction_block.Reaction_block</code></dt>
<dd>Reaction information</dd>
<dt><strong><code>EnzymeInfo</code></strong> :&ensp;<code>rbatools.enzyme_block.Enzyme_block</code></dt>
<dd>Enzyme information</dd>
<dt><strong><code>ProteinInfo</code></strong> :&ensp;<code>rbatools.protein_block.Protein_block</code></dt>
<dd>Protein information</dd>
<dt><strong><code>MacromoleculeInfo</code></strong> :&ensp;<code>rbatools.macromolecule_block.MacromoleculeBlock</code></dt>
<dd>Macromolecule (other than protein) information</dd>
<dt><strong><code>CompartmentInfo</code></strong> :&ensp;<code>rbatools.compartment_block.Compartment_block</code></dt>
<dd>Compartment information</dd>
<dt><strong><code>MetaboliteConstraintsInfo</code></strong> :&ensp;<code>rbatools.metabolite_constraints.Metabolite_constraints</code></dt>
<dd>Metabolite-constraint information</dd>
<dt><strong><code>DensityConstraintsInfo</code></strong> :&ensp;<code>rbatools.density_constraints.Density_constraints</code></dt>
<dd>Density-constraint information</dd>
<dt><strong><code>ProcessConstraintsInfo</code></strong> :&ensp;<code>rbatools.process_constraints.Process_constraints</code></dt>
<dd>Process-constraint information</dd>
<dt><strong><code>EnzymeConstraintsInfo</code></strong> :&ensp;<code>rbatools.enzyme_constraints.Enzyme_constraints</code></dt>
<dd>Enzyme-constraint information</dd>
<dt><strong><code>ModelStatistics</code></strong> :&ensp;<code>rbatools.statistics_block.Statistics_block</code></dt>
<dd>Statistics on Model</dd>
<dt><strong><code>TargetInfo</code></strong> :&ensp;<code>rbatools.target_block.TargetBlock</code></dt>
<dd>Target information</dd>
<dt><strong><code>ProteinMatrix</code></strong> :&ensp;<code>numpy.array</code></dt>
<dd>Matrix mapping proteins to consumers (enzymes and process-machineries)</dd>
<dt><strong><code>MediumDependencies</code></strong> :&ensp;<code>dict</code></dt>
<dd>Dictionary with boundary metabolites as keys
and a list of constraint_IDs, which are affected by its concentration, as values.</dd>
<dt><strong><code>MuDependencies</code></strong> :&ensp;<code>list</code></dt>
<dd>List of constraint_IDs, which are affected by the growth-rate</dd>
</dl>
<h2 id="methods">Methods</h2>
<p>fromFiles:
Generates model-structure object from model-files and provided auxilliary information.
fromJSON:
Generates model-structure object from model-structure in JSON-format.
exportJSON:
Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.
exportSBtab:
Saves model-structure object in SBtab-format under the specified name (filename).
generateMatrices:
Generates information-matricx object and stores it as attribute 'InfoMatrices'</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">class RBA_ModelStructure(object):
    &#34;&#34;&#34;
    Class holding information on model-structure.

    Attributes
    ----------
    GeneralInfo : rbatools.description_block.Description_block
         Model description
    MetaboliteInfo : rbatools.metabolite_block.Metabolite_block
         Metabolite information
    ModuleInfo : rbatools.module_block.Module_block
         Module information
    ProcessInfo : rbatools.process_block.Process_block
         Process information
    ReactionInfo : rbatools.reaction_block.Reaction_block
         Reaction information
    EnzymeInfo : rbatools.enzyme_block.Enzyme_block
         Enzyme information
    ProteinInfo : rbatools.protein_block.Protein_block
         Protein information
    MacromoleculeInfo : rbatools.macromolecule_block.MacromoleculeBlock
         Macromolecule (other than protein) information
    CompartmentInfo : rbatools.compartment_block.Compartment_block
         Compartment information
    MetaboliteConstraintsInfo : rbatools.metabolite_constraints.Metabolite_constraints
         Metabolite-constraint information
    DensityConstraintsInfo : rbatools.density_constraints.Density_constraints
         Density-constraint information
    ProcessConstraintsInfo : rbatools.process_constraints.Process_constraints
         Process-constraint information
    EnzymeConstraintsInfo : rbatools.enzyme_constraints.Enzyme_constraints
         Enzyme-constraint information
    ModelStatistics : rbatools.statistics_block.Statistics_block
         Statistics on Model
    TargetInfo : rbatools.target_block.TargetBlock
         Target information
    ProteinMatrix : numpy.array
        Matrix mapping proteins to consumers (enzymes and process-machineries)
    MediumDependencies : dict
        Dictionary with boundary metabolites as keys
        and a list of constraint_IDs, which are affected by its concentration, as values.
    MuDependencies : list
        List of constraint_IDs, which are affected by the growth-rate

    Methods
    ----------
    fromFiles:
        Generates model-structure object from model-files and provided auxilliary information.
    fromJSON:
        Generates model-structure object from model-structure in JSON-format.
    exportJSON:
        Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.
    exportSBtab:
        Saves model-structure object in SBtab-format under the specified name (filename).
    generateMatrices:
        Generates information-matricx object and stores it as attribute &#39;InfoMatrices&#39;
    &#34;&#34;&#34;

    def fromFiles(self, xml_dir):
        &#34;&#34;&#34;
        Generates model-structure object from model-files and provided auxilliary information.

        Parameters
        ----------
        xml_dir : str
            Directory, where RBA-model is located
        &#34;&#34;&#34;

        UniprotFile = importUniprotFile(xml_dir)
        GeneMap = importGeneAnnotations(xml_dir)
        Info = importModelInfo(xml_dir)
        SBMLfile = str(&#39;Not There&#39;)
        if Info[&#39;Value&#39;][&#39;SBML-file&#39;] != &#39;Not Provided&#39;:
            SBMLfile = importSbmlFile(xml_dir, str(Info[&#39;Value&#39;][&#39;SBML-file&#39;]))

        MetaboliteAnnotations = importMetaboliteAnnotations(xml_dir)
        ReactionAnnotations = importReactionAnnotations(xml_dir)

        print(&#39;&#39;)
        print(&#39;Generating model-structure&#39;)
        print(&#39;...&#39;)

        model = rba.RbaModel.from_xml(xml_dir)
        Zero_matrix = rba.ConstraintMatrix(model)
        Zero_matrix.build_matrices(0)
        constraints = sortConstraints(Zero_matrix, model)

        MetaboliteInfo = MetaboliteBlock()
        ModuleInfo = ModuleBlock()
        ProcessInfo = ProcessBlock()
        ReactionInfo = ReactionBlock()
        EnzymeInfo = EnzymeBlock()
        ProteinInfo = ProteinBlock()
        MacromoleculeInfo = MacromoleculeBlock()
        CompartmentInfo = CompartmentBlock()
        TargetInfo = TargetBlock()

        TargetInfo.fromFiles(model)
        MetaboliteInfo.fromFiles(model, Info, MetaboliteAnnotations, SBMLfile)
        ModuleInfo.fromFiles(model, SBMLfile)
        ProcessInfo.fromFiles(model, Info)
        ReactionInfo.fromFiles(model, Info, ReactionAnnotations, SBMLfile, MetaboliteInfo)
        EnzymeInfo.fromFiles(model, Info)
        ProteinInfo.fromFiles(model, GeneMap, Info, UniprotFile)
        MacromoleculeInfo.fromFiles(model)
        CompartmentInfo.fromFiles(model, Info)

        self.GeneralInfo = DescriptionBlock()
        self.GeneralInfo.fromFiles(Info)
        self.MetaboliteInfo = MetaboliteInfo
        self.ModuleInfo = ModuleInfo
        self.ProcessInfo = ProcessInfo
        self.TargetInfo = TargetInfo

        for protein in ProteinInfo.Elements.keys():
            AssociatedEnzyme = findAssociatedEnzyme(EnzymeInfo.Elements, protein)
            ProteinInfo.Elements[protein][&#39;associatedEnzymes&#39;] = AssociatedEnzyme[&#39;Enz&#39;]
            ProteinInfo.Elements[protein][&#39;associatedReactions&#39;] = AssociatedEnzyme[&#39;Rx&#39;]

        CB = ConstraintBlocks(model)
        for enzyme in EnzymeInfo.Elements.keys():
            EnzymeInfo.Elements[enzyme][&#39;Isozymes&#39;] = findIsozymes(
                enzyme, CB, ReactionInfo.Elements, EnzymeInfo.Elements[enzyme][&#39;Reaction&#39;])

        for rx in ReactionInfo.Elements.keys():
            if ReactionInfo.Elements[rx][&#39;Enzyme&#39;] is not &#39;&#39;:
                ReactionInfo.Elements[rx][&#39;Compartment_Machinery&#39;] = EnzymeInfo.Elements[ReactionInfo.Elements[rx]
                                                                                         [&#39;Enzyme&#39;]][&#39;EnzymeCompartment&#39;]

        for comp in CompartmentInfo.Elements.keys():
            ContainedMacromolecules = findContainedMacromolecules(comp, MacromoleculeInfo.Elements)
            ContainedProteins = findContainedProteins(comp, ProteinInfo.Elements)
            ContainedEnzymes = findContainedEnzymes(
                ContainedProteins, ProteinInfo.Elements, EnzymeInfo.Elements)
            CompartmentInfo.Elements[comp][&#39;associatedMacromolecules&#39;] = ContainedMacromolecules
            CompartmentInfo.Elements[comp][&#39;associatedProteins&#39;] = ContainedProteins
            CompartmentInfo.Elements[comp][&#39;associatedReactions&#39;] = ContainedEnzymes[&#39;containedReactions&#39;]
            CompartmentInfo.Elements[comp][&#39;associatedEnzymes&#39;] = ContainedEnzymes[&#39;containedEnzymes&#39;]

        self.ReactionInfo = ReactionInfo
        self.EnzymeInfo = EnzymeInfo
        self.ProteinInfo = ProteinInfo
        self.MacromoleculeInfo = MacromoleculeInfo
        self.CompartmentInfo = CompartmentInfo

        self.MetaboliteConstraintsInfo = MetaboliteConstraints()
        self.DensityConstraintsInfo = DensityConstraints()
        self.ProcessConstraintsInfo = ProcessConstraints()
        self.EnzymeConstraintsInfo = EnzymeConstraints()

        self.MetaboliteConstraintsInfo.fromFiles(constraints, Zero_matrix)
        self.DensityConstraintsInfo.fromFiles(model, constraints, Zero_matrix)
        self.ProcessConstraintsInfo.fromFiles(model, constraints, Zero_matrix)
        self.EnzymeConstraintsInfo.fromFiles(model, constraints, Zero_matrix)

        BioConstraintStats = StatsConstraintsBiological(self.MetaboliteConstraintsInfo.Elements,
                                                        self.EnzymeConstraintsInfo.Elements,
                                                        self.DensityConstraintsInfo.Elements,
                                                        self.ProcessConstraintsInfo.Elements)

        MathConstraintStats = StatsConstraintsMathematical(self.MetaboliteConstraintsInfo.Elements,
                                                           self.EnzymeConstraintsInfo.Elements,
                                                           self.DensityConstraintsInfo.Elements,
                                                           self.ProcessConstraintsInfo.Elements,
                                                           self.EnzymeInfo.Elements,
                                                           self.ReactionInfo.Elements,
                                                           self.ProcessInfo.Elements)

        FullOverview = generateOverview(self.ReactionInfo.overview(),
                                        self.MetaboliteInfo.overview(),
                                        self.ModuleInfo.overview(),
                                        self.EnzymeInfo.overview(),
                                        self.ProteinInfo.overview(),
                                        self.MacromoleculeInfo.overview(),
                                        self.ProcessInfo.overview(),
                                        self.CompartmentInfo.overview(),
                                        self.TargetInfo.overview(),
                                        BioConstraintStats,
                                        MathConstraintStats)

        self.ModelStatistics = StatisticsBlock()
        self.ModelStatistics.derive(FullOverview)

        self.ProteinMatrix = generateProteinMatrix(self)
        self.ProteinGeneMatrix = generateProteinGeneMatrix(self)
        self.MediumDependencies, self.MuDependencies = findParameterDependencies(self)

    def fromJSON(self, inputString):
        &#34;&#34;&#34;
        Generates model-structure object from model-structure in JSON-format.

        Parameters
        ----------
        inputString : json-str
            JSON-string to be parsed in to ModelStructure-object.
        &#34;&#34;&#34;
        Block = json.loads(inputString)
        self.ModelStatistics = StatisticsBlock()
        self.GeneralInfo = DescriptionBlock()
        self.ProcessInfo = ProcessBlock()
        self.CompartmentInfo = CompartmentBlock()
        self.MetaboliteInfo = MetaboliteBlock()
        self.TargetInfo = TargetBlock()
        self.ModuleInfo = ModuleBlock()
        self.EnzymeInfo = EnzymeBlock()
        self.ProteinInfo = ProteinBlock()
        self.MacromoleculeInfo = MacromoleculeBlock()
        self.ReactionInfo = ReactionBlock()
        self.DensityConstraintsInfo = DensityConstraints()
        self.ProcessConstraintsInfo = ProcessConstraints()
        self.MetaboliteConstraintsInfo = MetaboliteConstraints()
        self.EnzymeConstraintsInfo = EnzymeConstraints()

        self.ModelStatistics.fromDict(Block[&#39;ModelStatistics&#39;])
        self.GeneralInfo.fromDict(Block[&#39;ModelInformation&#39;])
        self.ProcessInfo.fromDict(Block[&#39;Process&#39;])
        self.CompartmentInfo.fromDict(Block[&#39;Compartment&#39;])
        self.MetaboliteInfo.fromDict(Block[&#39;Metabolite&#39;])
        self.ModuleInfo.fromDict(Block[&#39;Module&#39;])
        self.EnzymeInfo.fromDict(Block[&#39;Enzyme&#39;])
        self.ProteinInfo.fromDict(Block[&#39;Protein&#39;])
        self.MacromoleculeInfo.fromDict(Block[&#39;Macromolecule&#39;])
        self.ReactionInfo.fromDict(Block[&#39;Reaction&#39;])
        self.DensityConstraintsInfo.fromDict(Block[&#39;DensityConstraint&#39;])
        self.ProcessConstraintsInfo.fromDict(Block[&#39;ProcessConstraint&#39;])
        self.MetaboliteConstraintsInfo.fromDict(Block[&#39;MetaboliteConstraint&#39;])
        self.EnzymeConstraintsInfo.fromDict(Block[&#39;EnzymeConstraint&#39;])
        self.TargetInfo.fromDict(Block[&#39;Target&#39;])
        self.ProteinMatrix = Block[&#39;ProteinMatrix&#39;]
        self.ProteinMatrix[&#39;Matrix&#39;] = numpy.array(self.ProteinMatrix[&#39;Matrix&#39;])
        self.ProteinGeneMatrix = Block[&#39;ProteinGeneMatrix&#39;]
        self.ProteinGeneMatrix[&#39;Matrix&#39;] = numpy.array(self.ProteinGeneMatrix[&#39;Matrix&#39;])
        self.MediumDependencies = Block[&#39;MediumDependencies&#39;]
        self.MuDependencies = Block[&#39;MuDependencies&#39;]

    def exportJSON(self, path):
        &#34;&#34;&#34;
        Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.

        Parameters
        ----------
        path : str
            Directory, where to save JSON-file
        &#34;&#34;&#34;
        Block = {&#39;ModelInformation&#39;: self.GeneralInfo.Elements,
                 &#39;ModelStatistics&#39;: self.ModelStatistics.Elements,
                 &#39;Process&#39;: self.ProcessInfo.Elements,
                 &#39;Compartment&#39;: self.CompartmentInfo.Elements,
                 &#39;Metabolite&#39;: self.MetaboliteInfo.Elements,
                 &#39;Target&#39;: self.TargetInfo.Elements,
                 &#39;Module&#39;: self.ModuleInfo.Elements,
                 &#39;Enzyme&#39;: self.EnzymeInfo.Elements,
                 &#39;Protein&#39;: self.ProteinInfo.Elements,
                 &#39;Macromolecule&#39;: self.MacromoleculeInfo.Elements,
                 &#39;Reaction&#39;: self.ReactionInfo.Elements,
                 &#39;DensityConstraint&#39;: self.DensityConstraintsInfo.Elements,
                 &#39;ProcessConstraint&#39;: self.ProcessConstraintsInfo.Elements,
                 &#39;MetaboliteConstraint&#39;: self.MetaboliteConstraintsInfo.Elements,
                 &#39;EnzymeConstraint&#39;: self.EnzymeConstraintsInfo.Elements,
                 &#39;ProteinMatrix&#39;: self.ProteinMatrix,
                 &#39;ProteinGeneMatrix&#39;: self.ProteinGeneMatrix,
                 &#39;MediumDependencies&#39;: self.MediumDependencies,
                 &#39;MuDependencies&#39;: self.MuDependencies}
        # blocks.metabolism.external, model.metabolism.reactions
        # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;S&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
        # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;Reactions&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
        # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;Metabolites&#39;] = [i.id for i in blocks.metabolism.external]
        Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
        Block[&#39;ProteinGeneMatrix&#39;][&#39;Matrix&#39;] = Block[&#39;ProteinGeneMatrix&#39;][&#39;Matrix&#39;].tolist()
        JSONstring = json.dumps(Block, default=JSON_Int64_compensation)
        filename = path + &#39;/ModelStructure.json&#39;
        f = open(filename, &#39;w&#39;)
        f.write(JSONstring)
        f.close()
        return(JSONstring)

    def exportSBtab(self, add_links=False, filename=None):
        &#34;&#34;&#34;
        Saves model-structure object in SBtab-format under the specified name (filename).

        Parameters
        ----------
        filename : str
            Name, under which to save SBtab-file
        add_links : str
            Wheter to implement entry-format, which allows links between table-elements.
        &#34;&#34;&#34;
        GeneralInfoTable = self.GeneralInfo.toSBtab(table_id=&#39;ModelMetadata&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Model Information&#39;, Col_list=[
                                                    &#39;Measure&#39;, &#39;Value&#39;], NameList=[&#39;Element&#39;, &#39;Value&#39;])
        GeneralInfoTable.filename = &#39;ModelMetadata.tsv&#39;
        GeneralInfoTable.change_attribute(&#39;Text&#39;, &#39;Model metadata&#39;)
        GeneralInfoTable.unset_attribute(&#39;Date&#39;)
        GeneralInfoTable.unset_attribute(&#39;SBtabVersion&#39;)

        StatsTable = self.ModelStatistics.toSBtab(
            table_id=&#39;ModelElements&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Model Size&#39;, Col_list=[&#39;Measure&#39;, &#39;Value&#39;], NameList=[&#39;Element&#39;, &#39;Number&#39;])
        StatsTable.filename = &#39;ModelElements.tsv&#39;
        StatsTable.change_attribute(&#39;Text&#39;, &#39;Numbers of cell elements and constraints&#39;)
        StatsTable.unset_attribute(&#39;Date&#39;)
        StatsTable.unset_attribute(&#39;SBtabVersion&#39;)

        MetaboliteBlock_forChanges = copy.deepcopy(self.MetaboliteInfo)
        if add_links:
            for k in list(MetaboliteBlock_forChanges.Elements.keys()):
                if MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;] in self.CompartmentInfo.Elements.keys():
                    MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + &#39;Compartment&#39; + \
                        &#39;/&#39;+MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;]+&#39;!)&#39;
                for index, id in enumerate(MetaboliteBlock_forChanges.Elements[k][&#39;ReactionsInvolvedWith&#39;]):
                    MetaboliteBlock_forChanges.Elements[k][&#39;ReactionsInvolvedWith&#39;][
                        index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for id in list(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                    if not pandas.isna(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id]):
                        MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id] = str(&#39;(!identifiers:&#39;+id+&#39;/&#39;+str(
                            MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;|&#39;+str(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;!)&#39;)

        MetaboliteTable = MetaboliteBlock_forChanges.toSBtab(table_id=&#39;Compound&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Metabolites&#39;, Col_list=[
            &#39;ID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;ReactionsInvolvedWith&#39;, &#39;OtherIDs&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;Reactions&#39;, &#39;Annotation&#39;])
        MetaboliteTable.filename = &#39;Compound.tsv&#39;
        MetaboliteTable.change_attribute(
            &#39;Text&#39;, &#39;Metabolite species are localised in cell compartments and are associated with metabolite mass balance constraints.&#39;)
        MetaboliteTable.unset_attribute(&#39;Date&#39;)
        MetaboliteTable.unset_attribute(&#39;SBtabVersion&#39;)

        ReactionBlock_forChanges = copy.deepcopy(self.ReactionInfo)
        if add_links:
            for k in list(ReactionBlock_forChanges.Elements.keys()):
                oldEnz = ReactionBlock_forChanges.Elements[k][&#39;Enzyme&#39;]
                if len(oldEnz) &gt; 0:
                    ReactionBlock_forChanges.Elements[k][&#39;Enzyme&#39;] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+oldEnz+&#39;!)&#39;

                replacements = {}
                for i in ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;].split(&#39; &#39;):
                    if i in self.MetaboliteInfo.Elements.keys():
                        replacements[i] = &#39;(!&#39;+&#39;Compound&#39; + &#39;/&#39;+i+&#39;!)&#39;
                for i in replacements.keys():
                    ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;].replace(
                        str(&#39; &#39;+i+&#39; &#39;), str(&#39; &#39;+replacements[i]+&#39; &#39;))

                for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Compartment_Species&#39;]):
                    if id in self.CompartmentInfo.Elements.keys():
                        ReactionBlock_forChanges.Elements[k][&#39;Compartment_Species&#39;][
                            index] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;

                for id in list(ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;].keys()):
                    ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;][&#39;(!&#39;+&#39;Compound&#39; +
                                                                      &#39;/&#39;+id+&#39;!)&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;].pop(id)
                for id in list(ReactionBlock_forChanges.Elements[k][&#39;Products&#39;].keys()):
                    ReactionBlock_forChanges.Elements[k][&#39;Products&#39;][&#39;(!&#39;+&#39;Compound&#39; +
                                                                     &#39;/&#39;+id+&#39;!)&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Products&#39;].pop(id)
                for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Twins&#39;]):
                    ReactionBlock_forChanges.Elements[k][&#39;Twins&#39;][index] = &#39;(!&#39; + \
                        &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Compartment_Machinery&#39;]):
                    ReactionBlock_forChanges.Elements[k][&#39;Compartment_Machinery&#39;][
                        index] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;
                if &#39;ProtoID&#39; in list(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                    ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].pop(&#39;ProtoID&#39;)
                for id in list(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                    if not pandas.isna(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id]):
                        ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id] = str(&#39;(!identifiers:&#39;+id+&#39;/&#39;+str(
                            ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;|&#39;+str(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;!)&#39;)

        ReactionTable = ReactionBlock_forChanges.toSBtab(table_id=&#39;Reaction&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Reactions&#39;, Col_list=[&#39;ID&#39;, &#39;Name&#39;, &#39;Type&#39;, &#39;Reversible&#39;, &#39;Formula&#39;, &#39;Enzyme&#39;, &#39;Compartment_Machinery&#39;, &#39;Twins&#39;, &#39;Compartment_Species&#39;, &#39;OtherIDs&#39;], NameList=[
            &#39;ID&#39;, &#39;Name&#39;, &#39;Type&#39;, &#39;IsReversible&#39;, &#39;ReactionFormula&#39;, &#39;Enzyme&#39;, &#39;EnzymeCompartment&#39;, &#39;IsoenzymeReactions&#39;, &#39;ReactionCompartment&#39;, &#39;Annotation&#39;])
        ReactionTable.filename = &#39;Reaction.tsv&#39;
        ReactionTable.change_attribute(
            &#39;Text&#39;, &#39;Chemical reactions are localised in cell compartments. All reactions are enzyme catalysed.&#39;)
        ReactionTable.unset_attribute(&#39;Date&#39;)
        ReactionTable.unset_attribute(&#39;SBtabVersion&#39;)

        EnzymeBlock_forChanges = copy.deepcopy(self.EnzymeInfo)
        if add_links:
            for k in list(EnzymeBlock_forChanges.Elements.keys()):
                oldRx = EnzymeBlock_forChanges.Elements[k][&#39;Reaction&#39;]
                EnzymeBlock_forChanges.Elements[k][&#39;Reaction&#39;] = &#39;(!&#39;+&#39;Reaction&#39;+&#39;/&#39;+oldRx+&#39;!)&#39;
                for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;Isozymes&#39;]):
                    EnzymeBlock_forChanges.Elements[k][&#39;Isozymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;IdenticalEnzymes&#39;]):
                    EnzymeBlock_forChanges.Elements[k][&#39;IdenticalEnzymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;EnzymeCompartment&#39;]):
                    EnzymeBlock_forChanges.Elements[k][&#39;EnzymeCompartment&#39;][index] = &#39;(!&#39; + \
                        &#39;Compartment&#39;+&#39;/&#39; + id+&#39;!)&#39;
                for id in list(EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;].keys()):
                    EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;][&#39;(!&#39;+&#39;Protein&#39;+&#39;/&#39; +
                                                                   id+&#39;!)&#39;] = EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;].pop(id)

        EnzymeTable = EnzymeBlock_forChanges.toSBtab(table_id=&#39;Enzyme&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Enzymes&#39;, Col_list=[
            &#39;ID&#39;, &#39;Reaction&#39;, &#39;IdenticalEnzymes&#39;, &#39;Subunits&#39;, &#39;EnzymeCompartment&#39;, &#39;OtherIDs&#39;, &#39;Isozymes&#39;], NameList=[&#39;ID&#39;, &#39;CatalysedReaction&#39;, &#39;OtherEnzymaticActivities&#39;, &#39;Subunits&#39;, &#39;Compartment&#39;, &#39;Annotation&#39;, &#39;Isoenzymes&#39;])
        EnzymeTable.filename = &#39;Enzyme.tsv&#39;
        EnzymeTable.change_attribute(
            &#39;Text&#39;, &#39;Enzymes are localised in cell compartments and catalyse specific reactions. To describe multi-functional enzyme, RBA uses multiple enzyme entries. Enzymes are associated with enzyme capacity constraints.&#39;)
        EnzymeTable.unset_attribute(&#39;Date&#39;)
        EnzymeTable.unset_attribute(&#39;SBtabVersion&#39;)

        ProteinBlock_forChanges = copy.deepcopy(self.ProteinInfo)
        if add_links:
            for k in list(ProteinBlock_forChanges.Elements.keys()):
                oldComp = ProteinBlock_forChanges.Elements[k][&#39;Compartment&#39;]
                ProteinBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + \
                    &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;
                for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;associatedReactions&#39;]):
                    ProteinBlock_forChanges.Elements[k][&#39;associatedReactions&#39;][
                        index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;]):
                    ProteinBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;]):
                    ProteinBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;][index] = &#39;(!&#39; + \
                        &#39;Process&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for id in list(ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].keys()):
                    ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;][
                        &#39;(!&#39;+&#39;Process&#39;+&#39;/&#39; + id+&#39;!)&#39;] = ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].pop(id)
                for id in list(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;].keys()):
                    if not pandas.isna(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id]):
                        if id == &#39;UniprotID&#39;:
                            ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] = str(&#39;(!identifiers:uniprot/&#39;+str(
                                ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;|&#39;+str(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;!)&#39;)
                        if id == &#39;ECnumber&#39;:
                            if ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] != &#39;nan&#39;:
                                ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] = str(&#39;(!identifiers:ec-code/&#39;+str(
                                    ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id]).split(&#39;EC &#39;)[1]+&#39;|&#39;+str(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;!)&#39;)

        ProteinTable = ProteinBlock_forChanges.toSBtab(table_id=&#39;Protein&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Proteins&#39;, Col_list=[&#39;ID&#39;, &#39;ProtoID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;SupportsProcess&#39;, &#39;associatedEnzymes&#39;, &#39;associatedReactions&#39;, &#39;AAnumber&#39;, &#39;ProcessRequirements&#39;, &#39;Weight&#39;, &#39;ExternalIDs&#39;, &#39;Function&#39;], NameList=[
            &#39;ID&#39;, &#39;ProtoID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;ContributesToProcess&#39;, &#39;EnzymaticActivity&#39;, &#39;CatalysedReaction&#39;, &#39;ChainLength&#39;, &#39;RequiresProcess&#39;, &#39;MolecularWeight&#39;, &#39;Annotation&#39;, &#39;Function&#39;])
        ProteinTable.filename = &#39;Protein.tsv&#39;
        ProteinTable.change_attribute(
            &#39;Text&#39;, &#39;Proteins are localised in cell compartments and can be (or be part of) metabolic enzymes.&#39;)
        ProteinTable.unset_attribute(&#39;Date&#39;)
        ProteinTable.unset_attribute(&#39;SBtabVersion&#39;)

        MacromoleculeBlock_forChanges = copy.deepcopy(self.MacromoleculeInfo)
        if add_links:
            for k in list(MacromoleculeBlock_forChanges.Elements.keys()):
                oldComp = MacromoleculeBlock_forChanges.Elements[k][&#39;Compartment&#39;]
                MacromoleculeBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + \
                    &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;
                for index, id in enumerate(MacromoleculeBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;]):
                    MacromoleculeBlock_forChanges.Elements[k][
                        &#39;SupportsProcess&#39;][index] = &#39;(!&#39; + &#39;Process&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for id in list(MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].keys()):
                    MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;][
                        &#39;(!&#39;+&#39;Process&#39;+&#39;/&#39; + id+&#39;!)&#39;] = MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].pop(id)

        MacromoleculeTable = MacromoleculeBlock_forChanges.toSBtab(table_id=&#39;Macromolecule&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Macromolecules&#39;, Col_list=[
                                                                   &#39;ID&#39;, &#39;ProtoID&#39;, &#39;Type&#39;, &#39;Compartment&#39;, &#39;SupportsProcess&#39;, &#39;ProcessRequirements&#39;], NameList=[&#39;ID&#39;, &#39;ProtoID&#39;, &#39;Type&#39;, &#39;Compartment&#39;, &#39;SupportedtedProcess&#39;, &#39;RequiresProcess&#39;])
        MacromoleculeTable.filename = &#39;Macromolecules.tsv&#39;
        MacromoleculeTable.change_attribute(
            &#39;Text&#39;, &#39;Macromolecule are localised in cell compartments and can be part of cellular machinery or serve another function inside the cell.&#39;)
        MacromoleculeTable.unset_attribute(&#39;Date&#39;)
        MacromoleculeTable.unset_attribute(&#39;SBtabVersion&#39;)

        CompartmentBlock_forChanges = copy.deepcopy(self.CompartmentInfo)
        if add_links:
            for k in list(CompartmentBlock_forChanges.Elements.keys()):
                for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedReactions&#39;]):
                    CompartmentBlock_forChanges.Elements[k][&#39;associatedReactions&#39;][
                        index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;]):
                    CompartmentBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;][index] = &#39;(!&#39; + \
                        &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
                for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedProteins&#39;]):
                    CompartmentBlock_forChanges.Elements[k][&#39;associatedProteins&#39;][
                        index] = &#39;(!&#39; + &#39;Protein&#39;+&#39;/&#39;+id+&#39;!)&#39;

        CompartmentTable = CompartmentBlock_forChanges.toSBtab(table_id=&#39;Compartment&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Compartments&#39;, Col_list=[
                                                               &#39;ID&#39;, &#39;associatedProteins&#39;, &#39;associatedEnzymes&#39;, &#39;associatedReactions&#39;], NameList=[&#39;ID&#39;, &#39;Proteins&#39;, &#39;Enzymes&#39;, &#39;Reactions&#39;])

        CompartmentTable.filename = &#39;Compartment.tsv&#39;
        CompartmentTable.change_attribute(
            &#39;Text&#39;, &#39;Cell compartments are used to describe the localisation of proteins, enzymes, and reactions and are associated with density constraints.&#39;)
        CompartmentTable.unset_attribute(&#39;Date&#39;)
        CompartmentTable.unset_attribute(&#39;SBtabVersion&#39;)

        ModuleBlock_forChanges = copy.deepcopy(self.ModuleInfo)
        if add_links:
            for k in list(ModuleBlock_forChanges.Elements.keys()):
                for index, id in enumerate(ModuleBlock_forChanges.Elements[k][&#39;Members&#39;]):
                    if id in self.ReactionInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                    if id in self.ProteinInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Protein&#39;+&#39;/&#39;+id+&#39;!)&#39;
                    if id in self.MetaboliteInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Compound&#39;+&#39;/&#39;+id+&#39;!)&#39;
                    if id in self.CompartmentInfo.Elements.keys():
                        ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                            &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;

        ModuleTable = ModuleBlock_forChanges.toSBtab(table_id=&#39;CellModule&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Modules&#39;, Col_list=[
                                                     &#39;ID&#39;, &#39;Name&#39;, &#39;Members&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;Contains&#39;])
        ModuleTable.filename = &#39;Module.tsv&#39;
        ModuleTable.change_attribute(&#39;Text&#39;, &#39;Information on Modules in RBAmodel&#39;)
        ModuleTable.unset_attribute(&#39;Date&#39;)
        ModuleTable.unset_attribute(&#39;SBtabVersion&#39;)

        ProcessBlock_forChanges = copy.deepcopy(self.ProcessInfo)
        if add_links:
            for k in list(ProcessBlock_forChanges.Elements.keys()):
                for component in ProcessBlock_forChanges.Elements[k][&#39;Components&#39;].keys():
                    ProductKeys = list(
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;].keys())
                    for species in ProductKeys:
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;][
                            &#39;(!&#39;+&#39;Compound&#39;+&#39;/&#39;+species+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;].pop(species)
                    ReactantKeys = list(
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;].keys())
                    for species in ReactantKeys:
                        ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;][
                            &#39;(!&#39;+&#39;Compound&#39;+&#39;/&#39;+species+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;].pop(species)
                for id in list(ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].keys()):
                    if id in self.ProteinInfo.Elements.keys():
                        ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Protein&#39;+&#39;/&#39; +
                                                                           id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)
                    elif id in self.MacromoleculeInfo.Elements.keys():
                        ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Macromolecule&#39; +
                                                                           &#39;/&#39; + id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)
                    elif id in self.MetaboliteInfo.Elements.keys():
                        ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Metabolite&#39;+&#39;/&#39; +
                                                                           id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)

        ProcessTable = ProcessBlock_forChanges.toSBtab(table_id=&#39;Process&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Macromolecular Processes&#39;, Col_list=[
            &#39;ID&#39;, &#39;Name&#39;, &#39;Composition&#39;, &#39;Components&#39;, &#39;Initiation&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;MachineSubunits&#39;, &#39;MachineComponents&#39;, &#39;InitiationCofactors&#39;])
        ProcessTable.filename = &#39;Process.tsv&#39;
        ProcessTable.change_attribute(
            &#39;Text&#39;, &#39;Macromolecular machines catalyse the biochemical reactions that produce, modify, and degrade macromolecules. They are catalysed by macromolecular machines and associated with process capacity constraints.&#39;)
        ProcessTable.unset_attribute(&#39;Date&#39;)
        ProcessTable.unset_attribute(&#39;SBtabVersion&#39;)

        MetaboliteConstraintsBlock_forChanges = copy.deepcopy(self.MetaboliteConstraintsInfo)
        if add_links:
            for k in list(MetaboliteConstraintsBlock_forChanges.Elements.keys()):
                oldMet = MetaboliteConstraintsBlock_forChanges.Elements[k][&#39;AssociatedMetabolite&#39;]
                MetaboliteConstraintsBlock_forChanges.Elements[k][
                    &#39;AssociatedMetabolite&#39;] = &#39;(!&#39; + &#39;Compound&#39;+&#39;/&#39;+oldMet+&#39;!)&#39;

        MetaboliteConstraintTable = MetaboliteConstraintsBlock_forChanges.toSBtab(table_id=&#39;MetaboliteConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Metabolite Mass Balance Constraints&#39;, Col_list=[
                                                                                  &#39;ID&#39;, &#39;AssociatedMetabolite&#39;, &#39;Type&#39;], NameList=[&#39;ID&#39;, &#39;Metabolite&#39;, &#39;Type&#39;])
        MetaboliteConstraintTable.filename = &#39;MetaboliteConstraint.tsv&#39;
        MetaboliteConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;Metabolite mass balance constraints ensure mass balance and stationary fluxes in metabolism.&#39;)
        MetaboliteConstraintTable.unset_attribute(&#39;Date&#39;)
        MetaboliteConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        CompartmentConstraintsBlock_forChanges = copy.deepcopy(self.DensityConstraintsInfo)
        if add_links:
            for k in list(CompartmentConstraintsBlock_forChanges.Elements.keys()):
                oldComp = CompartmentConstraintsBlock_forChanges.Elements[k][&#39;AssociatedCompartment&#39;]
                CompartmentConstraintsBlock_forChanges.Elements[k][
                    &#39;AssociatedCompartment&#39;] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

        DensityConstraintTable = CompartmentConstraintsBlock_forChanges.toSBtab(table_id=&#39;DensityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Density Constraints&#39;, Col_list=[
                                                                                &#39;ID&#39;, &#39;AssociatedCompartment&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;Formula&#39;])
        DensityConstraintTable.filename = &#39;DensityConstraint.tsv&#39;
        DensityConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;Density constraints put an upper bound on the sum of macromolecule concentrations in a given compartment. The capacity parameter defines this bound in units corresponding to amino acids (contained in proteins), or one third of nucleotides (contained in RNA).&#39;)
        DensityConstraintTable.unset_attribute(&#39;Date&#39;)
        DensityConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        ProcessConstraintsBlock_forChanges = copy.deepcopy(self.ProcessConstraintsInfo)
        if add_links:
            for k in list(ProcessConstraintsBlock_forChanges.Elements.keys()):
                oldComp = ProcessConstraintsBlock_forChanges.Elements[k][&#39;AssociatedProcess&#39;]
                ProcessConstraintsBlock_forChanges.Elements[k][&#39;AssociatedProcess&#39;] = &#39;(!&#39; + \
                    &#39;Process&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

        ProcessConstraintTable = ProcessConstraintsBlock_forChanges.toSBtab(table_id=&#39;MachineryCapacityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Machinery Capacity Constraints&#39;, Col_list=[
            &#39;ID&#39;, &#39;AssociatedProcess&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Process&#39;, &#39;Type&#39;, &#39;Formula&#39;])
        ProcessConstraintTable.filename = &#39;MachineryConstraint.tsv&#39;
        ProcessConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;A machinery capacity constraint states that the rate of a macromolecular process is proportional to the concentration of the catalysing machine. The proportionality constant  (capacity parameter) can be a constant or a function of model parameters such as the growth rate.&#39;)
        ProcessConstraintTable.unset_attribute(&#39;Date&#39;)
        ProcessConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        EnzymeConstraintsBlock_forChanges = copy.deepcopy(self.EnzymeConstraintsInfo)
        if add_links:
            for k in list(EnzymeConstraintsBlock_forChanges.Elements.keys()):
                oldComp = EnzymeConstraintsBlock_forChanges.Elements[k][&#39;AssociatedEnzyme&#39;]
                EnzymeConstraintsBlock_forChanges.Elements[k][&#39;AssociatedEnzyme&#39;] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

        EnzymeConstraintTable = EnzymeConstraintsBlock_forChanges.toSBtab(table_id=&#39;EnzymeCapacityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Enzyme Capacity Constraints&#39;, Col_list=[
            &#39;ID&#39;, &#39;AssociatedEnzyme&#39;, &#39;AssociatedReaction&#39;, &#39;Direction&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Enzyme&#39;, &#39;Reaction&#39;, &#39;Direction&#39;, &#39;Type&#39;, &#39;Formula&#39;])
        EnzymeConstraintTable.filename = &#39;EnzymeConstraint.tsv&#39;
        EnzymeConstraintTable.change_attribute(
            &#39;Text&#39;, &#39;An enzyme capacity constraint states that a reaction rate is proportional to the concentration of the catalysing enzyme. The proportionality constant (capacity parameter) can be a constant or a function of model parameters such as the growth rate.&#39;)
        EnzymeConstraintTable.unset_attribute(&#39;Date&#39;)
        EnzymeConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

        TargetBlock_forChanges = copy.deepcopy(self.TargetInfo)
        if add_links:
            for k in list(TargetBlock_forChanges.Elements.keys()):
                oldTargSpec = TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;]
                if oldTargSpec in list(self.MetaboliteInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Compound&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
                if oldTargSpec in list(self.ReactionInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Reaction&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
                if oldTargSpec in list(self.ProteinInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Protein&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
                if oldTargSpec in list(self.MacromoleculeInfo.Elements.keys()):
                    TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                        &#39;Macromolecule&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;

        TargetTable = TargetBlock_forChanges.toSBtab(table_id=&#39;CellTarget&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Targets&#39;, Col_list=[
            &#39;ID&#39;, &#39;Group&#39;, &#39;Type&#39;, &#39;TargetEntity&#39;, &#39;TargetValue&#39;], NameList=[&#39;ID&#39;, &#39;TargetGroup&#39;, &#39;TargetType&#39;, &#39;TargetEntity&#39;, &#39;TargetParameter&#39;])
        TargetTable.filename = &#39;Target.tsv&#39;
        TargetTable.change_attribute(
            &#39;Text&#39;, &#39;Cell targets are biochemical variables (fluxes or concentrations) that need to remain below or above a certain threshold value to ensure a viable cell. They define constraints in the RBA model.&#39;)
        TargetTable.unset_attribute(&#39;Date&#39;)
        TargetTable.unset_attribute(&#39;SBtabVersion&#39;)

        if filename is not None:
            filename_SBtab = filename
        else:
            if add_links:
                filename_SBtab = &#39;RBA_model_withLinks&#39;
            else:
                filename_SBtab = &#39;RBA_model&#39;

        Out = SBtab.SBtabDocument(name=&#39;rbatools_withLinks&#39;, sbtab_init=None,
                                  filename=str(filename_SBtab+&#39;.tsv&#39;))
        Out.add_sbtab(GeneralInfoTable)
        Out.add_sbtab(StatsTable)
        Out.add_sbtab(MetaboliteTable)
        Out.add_sbtab(ReactionTable)
        Out.add_sbtab(EnzymeTable)
        Out.add_sbtab(ProteinTable)
        Out.add_sbtab(MacromoleculeTable)
        Out.add_sbtab(CompartmentTable)
        Out.add_sbtab(ModuleTable)
        Out.add_sbtab(ProcessTable)
        Out.add_sbtab(TargetTable)
        Out.add_sbtab(MetaboliteConstraintTable)
        Out.add_sbtab(DensityConstraintTable)
        Out.add_sbtab(ProcessConstraintTable)
        Out.add_sbtab(EnzymeConstraintTable)

        Out.name = filename_SBtab
        Out.change_attribute(&#39;DocumentName&#39;, self.GeneralInfo.Elements[&#39;Name&#39;])
        Out.change_attribute(&#39;DocumentType&#39;, &#39;rba-model-synopsis&#39;)
        Out.write()

    def generateMatrices(self):
        &#34;&#34;&#34;
        Generates information-matricx object and stores it as attribute &#39;InfoMatrices&#39;
        &#34;&#34;&#34;
        self.InfoMatrices = InfoMatrices(self)</code></pre>
</details>
<h3>Methods</h3>
<dl>
<dt id="rbatools.rba_ModelStructure.RBA_ModelStructure.exportJSON"><code class="name flex">
<span>def <span class="ident">exportJSON</span></span>(<span>self, path)</span>
</code></dt>
<dd>
<section class="desc"><p>Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>path</code></strong> :&ensp;<code>str</code></dt>
<dd>Directory, where to save JSON-file</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def exportJSON(self, path):
    &#34;&#34;&#34;
    Saves model-structure object in JSON-format in specified directory, under the name ModelStructure.json.

    Parameters
    ----------
    path : str
        Directory, where to save JSON-file
    &#34;&#34;&#34;
    Block = {&#39;ModelInformation&#39;: self.GeneralInfo.Elements,
             &#39;ModelStatistics&#39;: self.ModelStatistics.Elements,
             &#39;Process&#39;: self.ProcessInfo.Elements,
             &#39;Compartment&#39;: self.CompartmentInfo.Elements,
             &#39;Metabolite&#39;: self.MetaboliteInfo.Elements,
             &#39;Target&#39;: self.TargetInfo.Elements,
             &#39;Module&#39;: self.ModuleInfo.Elements,
             &#39;Enzyme&#39;: self.EnzymeInfo.Elements,
             &#39;Protein&#39;: self.ProteinInfo.Elements,
             &#39;Macromolecule&#39;: self.MacromoleculeInfo.Elements,
             &#39;Reaction&#39;: self.ReactionInfo.Elements,
             &#39;DensityConstraint&#39;: self.DensityConstraintsInfo.Elements,
             &#39;ProcessConstraint&#39;: self.ProcessConstraintsInfo.Elements,
             &#39;MetaboliteConstraint&#39;: self.MetaboliteConstraintsInfo.Elements,
             &#39;EnzymeConstraint&#39;: self.EnzymeConstraintsInfo.Elements,
             &#39;ProteinMatrix&#39;: self.ProteinMatrix,
             &#39;ProteinGeneMatrix&#39;: self.ProteinGeneMatrix,
             &#39;MediumDependencies&#39;: self.MediumDependencies,
             &#39;MuDependencies&#39;: self.MuDependencies}
    # blocks.metabolism.external, model.metabolism.reactions
    # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;S&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
    # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;Reactions&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
    # Block[&#39;ExternalStoichiometryMatrix&#39;][&#39;Metabolites&#39;] = [i.id for i in blocks.metabolism.external]
    Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;] = Block[&#39;ProteinMatrix&#39;][&#39;Matrix&#39;].tolist()
    Block[&#39;ProteinGeneMatrix&#39;][&#39;Matrix&#39;] = Block[&#39;ProteinGeneMatrix&#39;][&#39;Matrix&#39;].tolist()
    JSONstring = json.dumps(Block, default=JSON_Int64_compensation)
    filename = path + &#39;/ModelStructure.json&#39;
    f = open(filename, &#39;w&#39;)
    f.write(JSONstring)
    f.close()
    return(JSONstring)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.RBA_ModelStructure.exportSBtab"><code class="name flex">
<span>def <span class="ident">exportSBtab</span></span>(<span>self, add_links=False, filename=None)</span>
</code></dt>
<dd>
<section class="desc"><p>Saves model-structure object in SBtab-format under the specified name (filename).</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>filename</code></strong> :&ensp;<code>str</code></dt>
<dd>Name, under which to save SBtab-file</dd>
<dt><strong><code>add_links</code></strong> :&ensp;<code>str</code></dt>
<dd>Wheter to implement entry-format, which allows links between table-elements.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def exportSBtab(self, add_links=False, filename=None):
    &#34;&#34;&#34;
    Saves model-structure object in SBtab-format under the specified name (filename).

    Parameters
    ----------
    filename : str
        Name, under which to save SBtab-file
    add_links : str
        Wheter to implement entry-format, which allows links between table-elements.
    &#34;&#34;&#34;
    GeneralInfoTable = self.GeneralInfo.toSBtab(table_id=&#39;ModelMetadata&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Model Information&#39;, Col_list=[
                                                &#39;Measure&#39;, &#39;Value&#39;], NameList=[&#39;Element&#39;, &#39;Value&#39;])
    GeneralInfoTable.filename = &#39;ModelMetadata.tsv&#39;
    GeneralInfoTable.change_attribute(&#39;Text&#39;, &#39;Model metadata&#39;)
    GeneralInfoTable.unset_attribute(&#39;Date&#39;)
    GeneralInfoTable.unset_attribute(&#39;SBtabVersion&#39;)

    StatsTable = self.ModelStatistics.toSBtab(
        table_id=&#39;ModelElements&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Model Size&#39;, Col_list=[&#39;Measure&#39;, &#39;Value&#39;], NameList=[&#39;Element&#39;, &#39;Number&#39;])
    StatsTable.filename = &#39;ModelElements.tsv&#39;
    StatsTable.change_attribute(&#39;Text&#39;, &#39;Numbers of cell elements and constraints&#39;)
    StatsTable.unset_attribute(&#39;Date&#39;)
    StatsTable.unset_attribute(&#39;SBtabVersion&#39;)

    MetaboliteBlock_forChanges = copy.deepcopy(self.MetaboliteInfo)
    if add_links:
        for k in list(MetaboliteBlock_forChanges.Elements.keys()):
            if MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;] in self.CompartmentInfo.Elements.keys():
                MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + &#39;Compartment&#39; + \
                    &#39;/&#39;+MetaboliteBlock_forChanges.Elements[k][&#39;Compartment&#39;]+&#39;!)&#39;
            for index, id in enumerate(MetaboliteBlock_forChanges.Elements[k][&#39;ReactionsInvolvedWith&#39;]):
                MetaboliteBlock_forChanges.Elements[k][&#39;ReactionsInvolvedWith&#39;][
                    index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for id in list(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                if not pandas.isna(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id]):
                    MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id] = str(&#39;(!identifiers:&#39;+id+&#39;/&#39;+str(
                        MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;|&#39;+str(MetaboliteBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;!)&#39;)

    MetaboliteTable = MetaboliteBlock_forChanges.toSBtab(table_id=&#39;Compound&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Metabolites&#39;, Col_list=[
        &#39;ID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;ReactionsInvolvedWith&#39;, &#39;OtherIDs&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;Reactions&#39;, &#39;Annotation&#39;])
    MetaboliteTable.filename = &#39;Compound.tsv&#39;
    MetaboliteTable.change_attribute(
        &#39;Text&#39;, &#39;Metabolite species are localised in cell compartments and are associated with metabolite mass balance constraints.&#39;)
    MetaboliteTable.unset_attribute(&#39;Date&#39;)
    MetaboliteTable.unset_attribute(&#39;SBtabVersion&#39;)

    ReactionBlock_forChanges = copy.deepcopy(self.ReactionInfo)
    if add_links:
        for k in list(ReactionBlock_forChanges.Elements.keys()):
            oldEnz = ReactionBlock_forChanges.Elements[k][&#39;Enzyme&#39;]
            if len(oldEnz) &gt; 0:
                ReactionBlock_forChanges.Elements[k][&#39;Enzyme&#39;] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+oldEnz+&#39;!)&#39;

            replacements = {}
            for i in ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;].split(&#39; &#39;):
                if i in self.MetaboliteInfo.Elements.keys():
                    replacements[i] = &#39;(!&#39;+&#39;Compound&#39; + &#39;/&#39;+i+&#39;!)&#39;
            for i in replacements.keys():
                ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Formula&#39;].replace(
                    str(&#39; &#39;+i+&#39; &#39;), str(&#39; &#39;+replacements[i]+&#39; &#39;))

            for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Compartment_Species&#39;]):
                if id in self.CompartmentInfo.Elements.keys():
                    ReactionBlock_forChanges.Elements[k][&#39;Compartment_Species&#39;][
                        index] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;

            for id in list(ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;].keys()):
                ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;][&#39;(!&#39;+&#39;Compound&#39; +
                                                                  &#39;/&#39;+id+&#39;!)&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Reactants&#39;].pop(id)
            for id in list(ReactionBlock_forChanges.Elements[k][&#39;Products&#39;].keys()):
                ReactionBlock_forChanges.Elements[k][&#39;Products&#39;][&#39;(!&#39;+&#39;Compound&#39; +
                                                                 &#39;/&#39;+id+&#39;!)&#39;] = ReactionBlock_forChanges.Elements[k][&#39;Products&#39;].pop(id)
            for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Twins&#39;]):
                ReactionBlock_forChanges.Elements[k][&#39;Twins&#39;][index] = &#39;(!&#39; + \
                    &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(ReactionBlock_forChanges.Elements[k][&#39;Compartment_Machinery&#39;]):
                ReactionBlock_forChanges.Elements[k][&#39;Compartment_Machinery&#39;][
                    index] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;
            if &#39;ProtoID&#39; in list(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].pop(&#39;ProtoID&#39;)
            for id in list(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;].keys()):
                if not pandas.isna(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id]):
                    ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id] = str(&#39;(!identifiers:&#39;+id+&#39;/&#39;+str(
                        ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;|&#39;+str(ReactionBlock_forChanges.Elements[k][&#39;OtherIDs&#39;][id])+&#39;!)&#39;)

    ReactionTable = ReactionBlock_forChanges.toSBtab(table_id=&#39;Reaction&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Reactions&#39;, Col_list=[&#39;ID&#39;, &#39;Name&#39;, &#39;Type&#39;, &#39;Reversible&#39;, &#39;Formula&#39;, &#39;Enzyme&#39;, &#39;Compartment_Machinery&#39;, &#39;Twins&#39;, &#39;Compartment_Species&#39;, &#39;OtherIDs&#39;], NameList=[
        &#39;ID&#39;, &#39;Name&#39;, &#39;Type&#39;, &#39;IsReversible&#39;, &#39;ReactionFormula&#39;, &#39;Enzyme&#39;, &#39;EnzymeCompartment&#39;, &#39;IsoenzymeReactions&#39;, &#39;ReactionCompartment&#39;, &#39;Annotation&#39;])
    ReactionTable.filename = &#39;Reaction.tsv&#39;
    ReactionTable.change_attribute(
        &#39;Text&#39;, &#39;Chemical reactions are localised in cell compartments. All reactions are enzyme catalysed.&#39;)
    ReactionTable.unset_attribute(&#39;Date&#39;)
    ReactionTable.unset_attribute(&#39;SBtabVersion&#39;)

    EnzymeBlock_forChanges = copy.deepcopy(self.EnzymeInfo)
    if add_links:
        for k in list(EnzymeBlock_forChanges.Elements.keys()):
            oldRx = EnzymeBlock_forChanges.Elements[k][&#39;Reaction&#39;]
            EnzymeBlock_forChanges.Elements[k][&#39;Reaction&#39;] = &#39;(!&#39;+&#39;Reaction&#39;+&#39;/&#39;+oldRx+&#39;!)&#39;
            for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;Isozymes&#39;]):
                EnzymeBlock_forChanges.Elements[k][&#39;Isozymes&#39;][index] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;IdenticalEnzymes&#39;]):
                EnzymeBlock_forChanges.Elements[k][&#39;IdenticalEnzymes&#39;][index] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(EnzymeBlock_forChanges.Elements[k][&#39;EnzymeCompartment&#39;]):
                EnzymeBlock_forChanges.Elements[k][&#39;EnzymeCompartment&#39;][index] = &#39;(!&#39; + \
                    &#39;Compartment&#39;+&#39;/&#39; + id+&#39;!)&#39;
            for id in list(EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;].keys()):
                EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;][&#39;(!&#39;+&#39;Protein&#39;+&#39;/&#39; +
                                                               id+&#39;!)&#39;] = EnzymeBlock_forChanges.Elements[k][&#39;Subunits&#39;].pop(id)

    EnzymeTable = EnzymeBlock_forChanges.toSBtab(table_id=&#39;Enzyme&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Enzymes&#39;, Col_list=[
        &#39;ID&#39;, &#39;Reaction&#39;, &#39;IdenticalEnzymes&#39;, &#39;Subunits&#39;, &#39;EnzymeCompartment&#39;, &#39;OtherIDs&#39;, &#39;Isozymes&#39;], NameList=[&#39;ID&#39;, &#39;CatalysedReaction&#39;, &#39;OtherEnzymaticActivities&#39;, &#39;Subunits&#39;, &#39;Compartment&#39;, &#39;Annotation&#39;, &#39;Isoenzymes&#39;])
    EnzymeTable.filename = &#39;Enzyme.tsv&#39;
    EnzymeTable.change_attribute(
        &#39;Text&#39;, &#39;Enzymes are localised in cell compartments and catalyse specific reactions. To describe multi-functional enzyme, RBA uses multiple enzyme entries. Enzymes are associated with enzyme capacity constraints.&#39;)
    EnzymeTable.unset_attribute(&#39;Date&#39;)
    EnzymeTable.unset_attribute(&#39;SBtabVersion&#39;)

    ProteinBlock_forChanges = copy.deepcopy(self.ProteinInfo)
    if add_links:
        for k in list(ProteinBlock_forChanges.Elements.keys()):
            oldComp = ProteinBlock_forChanges.Elements[k][&#39;Compartment&#39;]
            ProteinBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + \
                &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;
            for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;associatedReactions&#39;]):
                ProteinBlock_forChanges.Elements[k][&#39;associatedReactions&#39;][
                    index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;]):
                ProteinBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;][index] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(ProteinBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;]):
                ProteinBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;][index] = &#39;(!&#39; + \
                    &#39;Process&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for id in list(ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].keys()):
                ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;][
                    &#39;(!&#39;+&#39;Process&#39;+&#39;/&#39; + id+&#39;!)&#39;] = ProteinBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].pop(id)
            for id in list(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;].keys()):
                if not pandas.isna(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id]):
                    if id == &#39;UniprotID&#39;:
                        ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] = str(&#39;(!identifiers:uniprot/&#39;+str(
                            ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;|&#39;+str(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;!)&#39;)
                    if id == &#39;ECnumber&#39;:
                        if ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] != &#39;nan&#39;:
                            ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id] = str(&#39;(!identifiers:ec-code/&#39;+str(
                                ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id]).split(&#39;EC &#39;)[1]+&#39;|&#39;+str(ProteinBlock_forChanges.Elements[k][&#39;ExternalIDs&#39;][id])+&#39;!)&#39;)

    ProteinTable = ProteinBlock_forChanges.toSBtab(table_id=&#39;Protein&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Proteins&#39;, Col_list=[&#39;ID&#39;, &#39;ProtoID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;SupportsProcess&#39;, &#39;associatedEnzymes&#39;, &#39;associatedReactions&#39;, &#39;AAnumber&#39;, &#39;ProcessRequirements&#39;, &#39;Weight&#39;, &#39;ExternalIDs&#39;, &#39;Function&#39;], NameList=[
        &#39;ID&#39;, &#39;ProtoID&#39;, &#39;Name&#39;, &#39;Compartment&#39;, &#39;ContributesToProcess&#39;, &#39;EnzymaticActivity&#39;, &#39;CatalysedReaction&#39;, &#39;ChainLength&#39;, &#39;RequiresProcess&#39;, &#39;MolecularWeight&#39;, &#39;Annotation&#39;, &#39;Function&#39;])
    ProteinTable.filename = &#39;Protein.tsv&#39;
    ProteinTable.change_attribute(
        &#39;Text&#39;, &#39;Proteins are localised in cell compartments and can be (or be part of) metabolic enzymes.&#39;)
    ProteinTable.unset_attribute(&#39;Date&#39;)
    ProteinTable.unset_attribute(&#39;SBtabVersion&#39;)

    MacromoleculeBlock_forChanges = copy.deepcopy(self.MacromoleculeInfo)
    if add_links:
        for k in list(MacromoleculeBlock_forChanges.Elements.keys()):
            oldComp = MacromoleculeBlock_forChanges.Elements[k][&#39;Compartment&#39;]
            MacromoleculeBlock_forChanges.Elements[k][&#39;Compartment&#39;] = &#39;(!&#39; + \
                &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;
            for index, id in enumerate(MacromoleculeBlock_forChanges.Elements[k][&#39;SupportsProcess&#39;]):
                MacromoleculeBlock_forChanges.Elements[k][
                    &#39;SupportsProcess&#39;][index] = &#39;(!&#39; + &#39;Process&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for id in list(MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].keys()):
                MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;][
                    &#39;(!&#39;+&#39;Process&#39;+&#39;/&#39; + id+&#39;!)&#39;] = MacromoleculeBlock_forChanges.Elements[k][&#39;ProcessRequirements&#39;].pop(id)

    MacromoleculeTable = MacromoleculeBlock_forChanges.toSBtab(table_id=&#39;Macromolecule&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Macromolecules&#39;, Col_list=[
                                                               &#39;ID&#39;, &#39;ProtoID&#39;, &#39;Type&#39;, &#39;Compartment&#39;, &#39;SupportsProcess&#39;, &#39;ProcessRequirements&#39;], NameList=[&#39;ID&#39;, &#39;ProtoID&#39;, &#39;Type&#39;, &#39;Compartment&#39;, &#39;SupportedtedProcess&#39;, &#39;RequiresProcess&#39;])
    MacromoleculeTable.filename = &#39;Macromolecules.tsv&#39;
    MacromoleculeTable.change_attribute(
        &#39;Text&#39;, &#39;Macromolecule are localised in cell compartments and can be part of cellular machinery or serve another function inside the cell.&#39;)
    MacromoleculeTable.unset_attribute(&#39;Date&#39;)
    MacromoleculeTable.unset_attribute(&#39;SBtabVersion&#39;)

    CompartmentBlock_forChanges = copy.deepcopy(self.CompartmentInfo)
    if add_links:
        for k in list(CompartmentBlock_forChanges.Elements.keys()):
            for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedReactions&#39;]):
                CompartmentBlock_forChanges.Elements[k][&#39;associatedReactions&#39;][
                    index] = &#39;(!&#39; + &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;]):
                CompartmentBlock_forChanges.Elements[k][&#39;associatedEnzymes&#39;][index] = &#39;(!&#39; + \
                    &#39;Enzyme&#39;+&#39;/&#39;+id+&#39;!)&#39;
            for index, id in enumerate(CompartmentBlock_forChanges.Elements[k][&#39;associatedProteins&#39;]):
                CompartmentBlock_forChanges.Elements[k][&#39;associatedProteins&#39;][
                    index] = &#39;(!&#39; + &#39;Protein&#39;+&#39;/&#39;+id+&#39;!)&#39;

    CompartmentTable = CompartmentBlock_forChanges.toSBtab(table_id=&#39;Compartment&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Compartments&#39;, Col_list=[
                                                           &#39;ID&#39;, &#39;associatedProteins&#39;, &#39;associatedEnzymes&#39;, &#39;associatedReactions&#39;], NameList=[&#39;ID&#39;, &#39;Proteins&#39;, &#39;Enzymes&#39;, &#39;Reactions&#39;])

    CompartmentTable.filename = &#39;Compartment.tsv&#39;
    CompartmentTable.change_attribute(
        &#39;Text&#39;, &#39;Cell compartments are used to describe the localisation of proteins, enzymes, and reactions and are associated with density constraints.&#39;)
    CompartmentTable.unset_attribute(&#39;Date&#39;)
    CompartmentTable.unset_attribute(&#39;SBtabVersion&#39;)

    ModuleBlock_forChanges = copy.deepcopy(self.ModuleInfo)
    if add_links:
        for k in list(ModuleBlock_forChanges.Elements.keys()):
            for index, id in enumerate(ModuleBlock_forChanges.Elements[k][&#39;Members&#39;]):
                if id in self.ReactionInfo.Elements.keys():
                    ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                        &#39;Reaction&#39;+&#39;/&#39;+id+&#39;!)&#39;
                if id in self.ProteinInfo.Elements.keys():
                    ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                        &#39;Protein&#39;+&#39;/&#39;+id+&#39;!)&#39;
                if id in self.MetaboliteInfo.Elements.keys():
                    ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                        &#39;Compound&#39;+&#39;/&#39;+id+&#39;!)&#39;
                if id in self.CompartmentInfo.Elements.keys():
                    ModuleBlock_forChanges.Elements[k][&#39;Members&#39;][index] = &#39;(!&#39; + \
                        &#39;Compartment&#39;+&#39;/&#39;+id+&#39;!)&#39;

    ModuleTable = ModuleBlock_forChanges.toSBtab(table_id=&#39;CellModule&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Modules&#39;, Col_list=[
                                                 &#39;ID&#39;, &#39;Name&#39;, &#39;Members&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;Contains&#39;])
    ModuleTable.filename = &#39;Module.tsv&#39;
    ModuleTable.change_attribute(&#39;Text&#39;, &#39;Information on Modules in RBAmodel&#39;)
    ModuleTable.unset_attribute(&#39;Date&#39;)
    ModuleTable.unset_attribute(&#39;SBtabVersion&#39;)

    ProcessBlock_forChanges = copy.deepcopy(self.ProcessInfo)
    if add_links:
        for k in list(ProcessBlock_forChanges.Elements.keys()):
            for component in ProcessBlock_forChanges.Elements[k][&#39;Components&#39;].keys():
                ProductKeys = list(
                    ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;].keys())
                for species in ProductKeys:
                    ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;][
                        &#39;(!&#39;+&#39;Compound&#39;+&#39;/&#39;+species+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Products&#39;].pop(species)
                ReactantKeys = list(
                    ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;].keys())
                for species in ReactantKeys:
                    ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;][
                        &#39;(!&#39;+&#39;Compound&#39;+&#39;/&#39;+species+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Components&#39;][component][&#39;Reactants&#39;].pop(species)
            for id in list(ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].keys()):
                if id in self.ProteinInfo.Elements.keys():
                    ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Protein&#39;+&#39;/&#39; +
                                                                       id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)
                elif id in self.MacromoleculeInfo.Elements.keys():
                    ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Macromolecule&#39; +
                                                                       &#39;/&#39; + id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)
                elif id in self.MetaboliteInfo.Elements.keys():
                    ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;][&#39;(!&#39;+&#39;Metabolite&#39;+&#39;/&#39; +
                                                                       id+&#39;!)&#39;] = ProcessBlock_forChanges.Elements[k][&#39;Composition&#39;].pop(id)

    ProcessTable = ProcessBlock_forChanges.toSBtab(table_id=&#39;Process&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Macromolecular Processes&#39;, Col_list=[
        &#39;ID&#39;, &#39;Name&#39;, &#39;Composition&#39;, &#39;Components&#39;, &#39;Initiation&#39;], NameList=[&#39;ID&#39;, &#39;Name&#39;, &#39;MachineSubunits&#39;, &#39;MachineComponents&#39;, &#39;InitiationCofactors&#39;])
    ProcessTable.filename = &#39;Process.tsv&#39;
    ProcessTable.change_attribute(
        &#39;Text&#39;, &#39;Macromolecular machines catalyse the biochemical reactions that produce, modify, and degrade macromolecules. They are catalysed by macromolecular machines and associated with process capacity constraints.&#39;)
    ProcessTable.unset_attribute(&#39;Date&#39;)
    ProcessTable.unset_attribute(&#39;SBtabVersion&#39;)

    MetaboliteConstraintsBlock_forChanges = copy.deepcopy(self.MetaboliteConstraintsInfo)
    if add_links:
        for k in list(MetaboliteConstraintsBlock_forChanges.Elements.keys()):
            oldMet = MetaboliteConstraintsBlock_forChanges.Elements[k][&#39;AssociatedMetabolite&#39;]
            MetaboliteConstraintsBlock_forChanges.Elements[k][
                &#39;AssociatedMetabolite&#39;] = &#39;(!&#39; + &#39;Compound&#39;+&#39;/&#39;+oldMet+&#39;!)&#39;

    MetaboliteConstraintTable = MetaboliteConstraintsBlock_forChanges.toSBtab(table_id=&#39;MetaboliteConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Metabolite Mass Balance Constraints&#39;, Col_list=[
                                                                              &#39;ID&#39;, &#39;AssociatedMetabolite&#39;, &#39;Type&#39;], NameList=[&#39;ID&#39;, &#39;Metabolite&#39;, &#39;Type&#39;])
    MetaboliteConstraintTable.filename = &#39;MetaboliteConstraint.tsv&#39;
    MetaboliteConstraintTable.change_attribute(
        &#39;Text&#39;, &#39;Metabolite mass balance constraints ensure mass balance and stationary fluxes in metabolism.&#39;)
    MetaboliteConstraintTable.unset_attribute(&#39;Date&#39;)
    MetaboliteConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

    CompartmentConstraintsBlock_forChanges = copy.deepcopy(self.DensityConstraintsInfo)
    if add_links:
        for k in list(CompartmentConstraintsBlock_forChanges.Elements.keys()):
            oldComp = CompartmentConstraintsBlock_forChanges.Elements[k][&#39;AssociatedCompartment&#39;]
            CompartmentConstraintsBlock_forChanges.Elements[k][
                &#39;AssociatedCompartment&#39;] = &#39;(!&#39; + &#39;Compartment&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

    DensityConstraintTable = CompartmentConstraintsBlock_forChanges.toSBtab(table_id=&#39;DensityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Density Constraints&#39;, Col_list=[
                                                                            &#39;ID&#39;, &#39;AssociatedCompartment&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Compartment&#39;, &#39;Type&#39;, &#39;Formula&#39;])
    DensityConstraintTable.filename = &#39;DensityConstraint.tsv&#39;
    DensityConstraintTable.change_attribute(
        &#39;Text&#39;, &#39;Density constraints put an upper bound on the sum of macromolecule concentrations in a given compartment. The capacity parameter defines this bound in units corresponding to amino acids (contained in proteins), or one third of nucleotides (contained in RNA).&#39;)
    DensityConstraintTable.unset_attribute(&#39;Date&#39;)
    DensityConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

    ProcessConstraintsBlock_forChanges = copy.deepcopy(self.ProcessConstraintsInfo)
    if add_links:
        for k in list(ProcessConstraintsBlock_forChanges.Elements.keys()):
            oldComp = ProcessConstraintsBlock_forChanges.Elements[k][&#39;AssociatedProcess&#39;]
            ProcessConstraintsBlock_forChanges.Elements[k][&#39;AssociatedProcess&#39;] = &#39;(!&#39; + \
                &#39;Process&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

    ProcessConstraintTable = ProcessConstraintsBlock_forChanges.toSBtab(table_id=&#39;MachineryCapacityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Machinery Capacity Constraints&#39;, Col_list=[
        &#39;ID&#39;, &#39;AssociatedProcess&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Process&#39;, &#39;Type&#39;, &#39;Formula&#39;])
    ProcessConstraintTable.filename = &#39;MachineryConstraint.tsv&#39;
    ProcessConstraintTable.change_attribute(
        &#39;Text&#39;, &#39;A machinery capacity constraint states that the rate of a macromolecular process is proportional to the concentration of the catalysing machine. The proportionality constant  (capacity parameter) can be a constant or a function of model parameters such as the growth rate.&#39;)
    ProcessConstraintTable.unset_attribute(&#39;Date&#39;)
    ProcessConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

    EnzymeConstraintsBlock_forChanges = copy.deepcopy(self.EnzymeConstraintsInfo)
    if add_links:
        for k in list(EnzymeConstraintsBlock_forChanges.Elements.keys()):
            oldComp = EnzymeConstraintsBlock_forChanges.Elements[k][&#39;AssociatedEnzyme&#39;]
            EnzymeConstraintsBlock_forChanges.Elements[k][&#39;AssociatedEnzyme&#39;] = &#39;(!&#39; + \
                &#39;Enzyme&#39;+&#39;/&#39;+oldComp+&#39;!)&#39;

    EnzymeConstraintTable = EnzymeConstraintsBlock_forChanges.toSBtab(table_id=&#39;EnzymeCapacityConstraint&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Enzyme Capacity Constraints&#39;, Col_list=[
        &#39;ID&#39;, &#39;AssociatedEnzyme&#39;, &#39;AssociatedReaction&#39;, &#39;Direction&#39;, &#39;Type&#39;, &#39;CapacityParameter&#39;], NameList=[&#39;ID&#39;, &#39;Enzyme&#39;, &#39;Reaction&#39;, &#39;Direction&#39;, &#39;Type&#39;, &#39;Formula&#39;])
    EnzymeConstraintTable.filename = &#39;EnzymeConstraint.tsv&#39;
    EnzymeConstraintTable.change_attribute(
        &#39;Text&#39;, &#39;An enzyme capacity constraint states that a reaction rate is proportional to the concentration of the catalysing enzyme. The proportionality constant (capacity parameter) can be a constant or a function of model parameters such as the growth rate.&#39;)
    EnzymeConstraintTable.unset_attribute(&#39;Date&#39;)
    EnzymeConstraintTable.unset_attribute(&#39;SBtabVersion&#39;)

    TargetBlock_forChanges = copy.deepcopy(self.TargetInfo)
    if add_links:
        for k in list(TargetBlock_forChanges.Elements.keys()):
            oldTargSpec = TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;]
            if oldTargSpec in list(self.MetaboliteInfo.Elements.keys()):
                TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                    &#39;Compound&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
            if oldTargSpec in list(self.ReactionInfo.Elements.keys()):
                TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                    &#39;Reaction&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
            if oldTargSpec in list(self.ProteinInfo.Elements.keys()):
                TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                    &#39;Protein&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;
            if oldTargSpec in list(self.MacromoleculeInfo.Elements.keys()):
                TargetBlock_forChanges.Elements[k][&#39;TargetEntity&#39;] = &#39;(!&#39; + \
                    &#39;Macromolecule&#39;+&#39;/&#39;+oldTargSpec+&#39;!)&#39;

    TargetTable = TargetBlock_forChanges.toSBtab(table_id=&#39;CellTarget&#39;, table_type=&#39;Quantity&#39;, table_name=&#39;Cell Targets&#39;, Col_list=[
        &#39;ID&#39;, &#39;Group&#39;, &#39;Type&#39;, &#39;TargetEntity&#39;, &#39;TargetValue&#39;], NameList=[&#39;ID&#39;, &#39;TargetGroup&#39;, &#39;TargetType&#39;, &#39;TargetEntity&#39;, &#39;TargetParameter&#39;])
    TargetTable.filename = &#39;Target.tsv&#39;
    TargetTable.change_attribute(
        &#39;Text&#39;, &#39;Cell targets are biochemical variables (fluxes or concentrations) that need to remain below or above a certain threshold value to ensure a viable cell. They define constraints in the RBA model.&#39;)
    TargetTable.unset_attribute(&#39;Date&#39;)
    TargetTable.unset_attribute(&#39;SBtabVersion&#39;)

    if filename is not None:
        filename_SBtab = filename
    else:
        if add_links:
            filename_SBtab = &#39;RBA_model_withLinks&#39;
        else:
            filename_SBtab = &#39;RBA_model&#39;

    Out = SBtab.SBtabDocument(name=&#39;rbatools_withLinks&#39;, sbtab_init=None,
                              filename=str(filename_SBtab+&#39;.tsv&#39;))
    Out.add_sbtab(GeneralInfoTable)
    Out.add_sbtab(StatsTable)
    Out.add_sbtab(MetaboliteTable)
    Out.add_sbtab(ReactionTable)
    Out.add_sbtab(EnzymeTable)
    Out.add_sbtab(ProteinTable)
    Out.add_sbtab(MacromoleculeTable)
    Out.add_sbtab(CompartmentTable)
    Out.add_sbtab(ModuleTable)
    Out.add_sbtab(ProcessTable)
    Out.add_sbtab(TargetTable)
    Out.add_sbtab(MetaboliteConstraintTable)
    Out.add_sbtab(DensityConstraintTable)
    Out.add_sbtab(ProcessConstraintTable)
    Out.add_sbtab(EnzymeConstraintTable)

    Out.name = filename_SBtab
    Out.change_attribute(&#39;DocumentName&#39;, self.GeneralInfo.Elements[&#39;Name&#39;])
    Out.change_attribute(&#39;DocumentType&#39;, &#39;rba-model-synopsis&#39;)
    Out.write()</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.RBA_ModelStructure.fromFiles"><code class="name flex">
<span>def <span class="ident">fromFiles</span></span>(<span>self, xml_dir)</span>
</code></dt>
<dd>
<section class="desc"><p>Generates model-structure object from model-files and provided auxilliary information.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>xml_dir</code></strong> :&ensp;<code>str</code></dt>
<dd>Directory, where RBA-model is located</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def fromFiles(self, xml_dir):
    &#34;&#34;&#34;
    Generates model-structure object from model-files and provided auxilliary information.

    Parameters
    ----------
    xml_dir : str
        Directory, where RBA-model is located
    &#34;&#34;&#34;

    UniprotFile = importUniprotFile(xml_dir)
    GeneMap = importGeneAnnotations(xml_dir)
    Info = importModelInfo(xml_dir)
    SBMLfile = str(&#39;Not There&#39;)
    if Info[&#39;Value&#39;][&#39;SBML-file&#39;] != &#39;Not Provided&#39;:
        SBMLfile = importSbmlFile(xml_dir, str(Info[&#39;Value&#39;][&#39;SBML-file&#39;]))

    MetaboliteAnnotations = importMetaboliteAnnotations(xml_dir)
    ReactionAnnotations = importReactionAnnotations(xml_dir)

    print(&#39;&#39;)
    print(&#39;Generating model-structure&#39;)
    print(&#39;...&#39;)

    model = rba.RbaModel.from_xml(xml_dir)
    Zero_matrix = rba.ConstraintMatrix(model)
    Zero_matrix.build_matrices(0)
    constraints = sortConstraints(Zero_matrix, model)

    MetaboliteInfo = MetaboliteBlock()
    ModuleInfo = ModuleBlock()
    ProcessInfo = ProcessBlock()
    ReactionInfo = ReactionBlock()
    EnzymeInfo = EnzymeBlock()
    ProteinInfo = ProteinBlock()
    MacromoleculeInfo = MacromoleculeBlock()
    CompartmentInfo = CompartmentBlock()
    TargetInfo = TargetBlock()

    TargetInfo.fromFiles(model)
    MetaboliteInfo.fromFiles(model, Info, MetaboliteAnnotations, SBMLfile)
    ModuleInfo.fromFiles(model, SBMLfile)
    ProcessInfo.fromFiles(model, Info)
    ReactionInfo.fromFiles(model, Info, ReactionAnnotations, SBMLfile, MetaboliteInfo)
    EnzymeInfo.fromFiles(model, Info)
    ProteinInfo.fromFiles(model, GeneMap, Info, UniprotFile)
    MacromoleculeInfo.fromFiles(model)
    CompartmentInfo.fromFiles(model, Info)

    self.GeneralInfo = DescriptionBlock()
    self.GeneralInfo.fromFiles(Info)
    self.MetaboliteInfo = MetaboliteInfo
    self.ModuleInfo = ModuleInfo
    self.ProcessInfo = ProcessInfo
    self.TargetInfo = TargetInfo

    for protein in ProteinInfo.Elements.keys():
        AssociatedEnzyme = findAssociatedEnzyme(EnzymeInfo.Elements, protein)
        ProteinInfo.Elements[protein][&#39;associatedEnzymes&#39;] = AssociatedEnzyme[&#39;Enz&#39;]
        ProteinInfo.Elements[protein][&#39;associatedReactions&#39;] = AssociatedEnzyme[&#39;Rx&#39;]

    CB = ConstraintBlocks(model)
    for enzyme in EnzymeInfo.Elements.keys():
        EnzymeInfo.Elements[enzyme][&#39;Isozymes&#39;] = findIsozymes(
            enzyme, CB, ReactionInfo.Elements, EnzymeInfo.Elements[enzyme][&#39;Reaction&#39;])

    for rx in ReactionInfo.Elements.keys():
        if ReactionInfo.Elements[rx][&#39;Enzyme&#39;] is not &#39;&#39;:
            ReactionInfo.Elements[rx][&#39;Compartment_Machinery&#39;] = EnzymeInfo.Elements[ReactionInfo.Elements[rx]
                                                                                     [&#39;Enzyme&#39;]][&#39;EnzymeCompartment&#39;]

    for comp in CompartmentInfo.Elements.keys():
        ContainedMacromolecules = findContainedMacromolecules(comp, MacromoleculeInfo.Elements)
        ContainedProteins = findContainedProteins(comp, ProteinInfo.Elements)
        ContainedEnzymes = findContainedEnzymes(
            ContainedProteins, ProteinInfo.Elements, EnzymeInfo.Elements)
        CompartmentInfo.Elements[comp][&#39;associatedMacromolecules&#39;] = ContainedMacromolecules
        CompartmentInfo.Elements[comp][&#39;associatedProteins&#39;] = ContainedProteins
        CompartmentInfo.Elements[comp][&#39;associatedReactions&#39;] = ContainedEnzymes[&#39;containedReactions&#39;]
        CompartmentInfo.Elements[comp][&#39;associatedEnzymes&#39;] = ContainedEnzymes[&#39;containedEnzymes&#39;]

    self.ReactionInfo = ReactionInfo
    self.EnzymeInfo = EnzymeInfo
    self.ProteinInfo = ProteinInfo
    self.MacromoleculeInfo = MacromoleculeInfo
    self.CompartmentInfo = CompartmentInfo

    self.MetaboliteConstraintsInfo = MetaboliteConstraints()
    self.DensityConstraintsInfo = DensityConstraints()
    self.ProcessConstraintsInfo = ProcessConstraints()
    self.EnzymeConstraintsInfo = EnzymeConstraints()

    self.MetaboliteConstraintsInfo.fromFiles(constraints, Zero_matrix)
    self.DensityConstraintsInfo.fromFiles(model, constraints, Zero_matrix)
    self.ProcessConstraintsInfo.fromFiles(model, constraints, Zero_matrix)
    self.EnzymeConstraintsInfo.fromFiles(model, constraints, Zero_matrix)

    BioConstraintStats = StatsConstraintsBiological(self.MetaboliteConstraintsInfo.Elements,
                                                    self.EnzymeConstraintsInfo.Elements,
                                                    self.DensityConstraintsInfo.Elements,
                                                    self.ProcessConstraintsInfo.Elements)

    MathConstraintStats = StatsConstraintsMathematical(self.MetaboliteConstraintsInfo.Elements,
                                                       self.EnzymeConstraintsInfo.Elements,
                                                       self.DensityConstraintsInfo.Elements,
                                                       self.ProcessConstraintsInfo.Elements,
                                                       self.EnzymeInfo.Elements,
                                                       self.ReactionInfo.Elements,
                                                       self.ProcessInfo.Elements)

    FullOverview = generateOverview(self.ReactionInfo.overview(),
                                    self.MetaboliteInfo.overview(),
                                    self.ModuleInfo.overview(),
                                    self.EnzymeInfo.overview(),
                                    self.ProteinInfo.overview(),
                                    self.MacromoleculeInfo.overview(),
                                    self.ProcessInfo.overview(),
                                    self.CompartmentInfo.overview(),
                                    self.TargetInfo.overview(),
                                    BioConstraintStats,
                                    MathConstraintStats)

    self.ModelStatistics = StatisticsBlock()
    self.ModelStatistics.derive(FullOverview)

    self.ProteinMatrix = generateProteinMatrix(self)
    self.ProteinGeneMatrix = generateProteinGeneMatrix(self)
    self.MediumDependencies, self.MuDependencies = findParameterDependencies(self)</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.RBA_ModelStructure.fromJSON"><code class="name flex">
<span>def <span class="ident">fromJSON</span></span>(<span>self, inputString)</span>
</code></dt>
<dd>
<section class="desc"><p>Generates model-structure object from model-structure in JSON-format.</p>
<h2 id="parameters">Parameters</h2>
<dl>
<dt><strong><code>inputString</code></strong> :&ensp;<code>json</code>-<code>str</code></dt>
<dd>JSON-string to be parsed in to ModelStructure-object.</dd>
</dl></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def fromJSON(self, inputString):
    &#34;&#34;&#34;
    Generates model-structure object from model-structure in JSON-format.

    Parameters
    ----------
    inputString : json-str
        JSON-string to be parsed in to ModelStructure-object.
    &#34;&#34;&#34;
    Block = json.loads(inputString)
    self.ModelStatistics = StatisticsBlock()
    self.GeneralInfo = DescriptionBlock()
    self.ProcessInfo = ProcessBlock()
    self.CompartmentInfo = CompartmentBlock()
    self.MetaboliteInfo = MetaboliteBlock()
    self.TargetInfo = TargetBlock()
    self.ModuleInfo = ModuleBlock()
    self.EnzymeInfo = EnzymeBlock()
    self.ProteinInfo = ProteinBlock()
    self.MacromoleculeInfo = MacromoleculeBlock()
    self.ReactionInfo = ReactionBlock()
    self.DensityConstraintsInfo = DensityConstraints()
    self.ProcessConstraintsInfo = ProcessConstraints()
    self.MetaboliteConstraintsInfo = MetaboliteConstraints()
    self.EnzymeConstraintsInfo = EnzymeConstraints()

    self.ModelStatistics.fromDict(Block[&#39;ModelStatistics&#39;])
    self.GeneralInfo.fromDict(Block[&#39;ModelInformation&#39;])
    self.ProcessInfo.fromDict(Block[&#39;Process&#39;])
    self.CompartmentInfo.fromDict(Block[&#39;Compartment&#39;])
    self.MetaboliteInfo.fromDict(Block[&#39;Metabolite&#39;])
    self.ModuleInfo.fromDict(Block[&#39;Module&#39;])
    self.EnzymeInfo.fromDict(Block[&#39;Enzyme&#39;])
    self.ProteinInfo.fromDict(Block[&#39;Protein&#39;])
    self.MacromoleculeInfo.fromDict(Block[&#39;Macromolecule&#39;])
    self.ReactionInfo.fromDict(Block[&#39;Reaction&#39;])
    self.DensityConstraintsInfo.fromDict(Block[&#39;DensityConstraint&#39;])
    self.ProcessConstraintsInfo.fromDict(Block[&#39;ProcessConstraint&#39;])
    self.MetaboliteConstraintsInfo.fromDict(Block[&#39;MetaboliteConstraint&#39;])
    self.EnzymeConstraintsInfo.fromDict(Block[&#39;EnzymeConstraint&#39;])
    self.TargetInfo.fromDict(Block[&#39;Target&#39;])
    self.ProteinMatrix = Block[&#39;ProteinMatrix&#39;]
    self.ProteinMatrix[&#39;Matrix&#39;] = numpy.array(self.ProteinMatrix[&#39;Matrix&#39;])
    self.ProteinGeneMatrix = Block[&#39;ProteinGeneMatrix&#39;]
    self.ProteinGeneMatrix[&#39;Matrix&#39;] = numpy.array(self.ProteinGeneMatrix[&#39;Matrix&#39;])
    self.MediumDependencies = Block[&#39;MediumDependencies&#39;]
    self.MuDependencies = Block[&#39;MuDependencies&#39;]</code></pre>
</details>
</dd>
<dt id="rbatools.rba_ModelStructure.RBA_ModelStructure.generateMatrices"><code class="name flex">
<span>def <span class="ident">generateMatrices</span></span>(<span>self)</span>
</code></dt>
<dd>
<section class="desc"><p>Generates information-matricx object and stores it as attribute 'InfoMatrices'</p></section>
<details class="source">
<summary>Source code</summary>
<pre><code class="python">def generateMatrices(self):
    &#34;&#34;&#34;
    Generates information-matricx object and stores it as attribute &#39;InfoMatrices&#39;
    &#34;&#34;&#34;
    self.InfoMatrices = InfoMatrices(self)</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="rbatools" href="index.html">rbatools</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="rbatools.rba_ModelStructure.JSON_Int64_compensation" href="#rbatools.rba_ModelStructure.JSON_Int64_compensation">JSON_Int64_compensation</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.StatsConstraintsBiological" href="#rbatools.rba_ModelStructure.StatsConstraintsBiological">StatsConstraintsBiological</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.StatsConstraintsMathematical" href="#rbatools.rba_ModelStructure.StatsConstraintsMathematical">StatsConstraintsMathematical</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.findAssociatedEnzyme" href="#rbatools.rba_ModelStructure.findAssociatedEnzyme">findAssociatedEnzyme</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.findContainedEnzymes" href="#rbatools.rba_ModelStructure.findContainedEnzymes">findContainedEnzymes</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.findContainedMacromolecules" href="#rbatools.rba_ModelStructure.findContainedMacromolecules">findContainedMacromolecules</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.findContainedProteins" href="#rbatools.rba_ModelStructure.findContainedProteins">findContainedProteins</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.findIsozymes" href="#rbatools.rba_ModelStructure.findIsozymes">findIsozymes</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.findParameterDependencies" href="#rbatools.rba_ModelStructure.findParameterDependencies">findParameterDependencies</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.generateOverview" href="#rbatools.rba_ModelStructure.generateOverview">generateOverview</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.generateProteinGeneMatrix" href="#rbatools.rba_ModelStructure.generateProteinGeneMatrix">generateProteinGeneMatrix</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.generateProteinMatrix" href="#rbatools.rba_ModelStructure.generateProteinMatrix">generateProteinMatrix</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.htmlStyle" href="#rbatools.rba_ModelStructure.htmlStyle">htmlStyle</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importBiggMetabolites" href="#rbatools.rba_ModelStructure.importBiggMetabolites">importBiggMetabolites</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importBiggReactions" href="#rbatools.rba_ModelStructure.importBiggReactions">importBiggReactions</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importGeneAnnotations" href="#rbatools.rba_ModelStructure.importGeneAnnotations">importGeneAnnotations</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importMetaboliteAnnotations" href="#rbatools.rba_ModelStructure.importMetaboliteAnnotations">importMetaboliteAnnotations</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importModelInfo" href="#rbatools.rba_ModelStructure.importModelInfo">importModelInfo</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importReactionAnnotations" href="#rbatools.rba_ModelStructure.importReactionAnnotations">importReactionAnnotations</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importSbmlFile" href="#rbatools.rba_ModelStructure.importSbmlFile">importSbmlFile</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.importUniprotFile" href="#rbatools.rba_ModelStructure.importUniprotFile">importUniprotFile</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.sortConstraints" href="#rbatools.rba_ModelStructure.sortConstraints">sortConstraints</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="rbatools.rba_ModelStructure.RBA_ModelStructure" href="#rbatools.rba_ModelStructure.RBA_ModelStructure">RBA_ModelStructure</a></code></h4>
<ul class="">
<li><code><a title="rbatools.rba_ModelStructure.RBA_ModelStructure.exportJSON" href="#rbatools.rba_ModelStructure.RBA_ModelStructure.exportJSON">exportJSON</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.RBA_ModelStructure.exportSBtab" href="#rbatools.rba_ModelStructure.RBA_ModelStructure.exportSBtab">exportSBtab</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.RBA_ModelStructure.fromFiles" href="#rbatools.rba_ModelStructure.RBA_ModelStructure.fromFiles">fromFiles</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.RBA_ModelStructure.fromJSON" href="#rbatools.rba_ModelStructure.RBA_ModelStructure.fromJSON">fromJSON</a></code></li>
<li><code><a title="rbatools.rba_ModelStructure.RBA_ModelStructure.generateMatrices" href="#rbatools.rba_ModelStructure.RBA_ModelStructure.generateMatrices">generateMatrices</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.5.3</a>.</p>
</footer>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
<script>hljs.initHighlightingOnLoad()</script>
</body>
</html>
